---
title: "hN project"
subtitle: "Zoom in immune cells"
author: "Audrey"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot"),
  time_it = local({
    now = NULL
    function(before, options) {
      if (options$time_it) {
        if (before) {
          now <= Sys.time()
        } else {
          res = difftime(Sys.time(), now, units = "secs")
          paste("(Time to run :", round(res, digits = 2), "s)")
        }
      }
    }
  })
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1337L)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # usefull for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15,
                      
                      # something about seed, chunk and Rmarkdown compilation
                      # https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script
                      # cache = TRUE,
                      cache.lazy = FALSE, 
                      
                      # add runtime after chunk
                      time_it = FALSE)
```


This file is used to analyse the immune cells dataset.

```{r library}
library(dplyr)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)
library(org.Mm.eg.db)

.libPaths()
```

# Preparation

In this section, we set the global settings of the analysis. We will store data there :

```{r out_dir}
save_name = "immune_cells"
out_dir = "."
```

We load the dataset :

```{r load_sobj}
sobj = readRDS(paste0(out_dir, "/", save_name, "_sobj.rds"))
sobj
```

and sample information :

```{r load_sample_info}
sample_info = readRDS(paste0(out_dir, "/", save_name, "_sample_info.rds")) 
sample_info
```

Here are custom colors for each cell type :

```{r color_markers, fig.width = 10, fig.height = 1, class.source = "fold-hide"}
color_markers = readRDS(paste0(out_dir, "/../../2_metadata/hN_metadata_color_markers.rds"))

data.frame(cell_type = names(color_markers),
           color = unlist(color_markers)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(color_markers), breaks = names(color_markers)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank())
```

This is the projection of interest :

```{r name2D}
name2D = "harmony_40_tsne"
```

We design a custom functions to represent cells of interest on the projection :

```{r see_clusters, class.source = "fold-hide"}
see_clusters = function(pop_oi = "DC") {
  # Clusters containing population of interest
  clusters_oi = names(which(cell_type_clusters == pop_oi))
  
  # Colors for clusters
  custom_colors = aquarius::gg_color_hue(length(levels(sobj$seurat_clusters)))
  names(custom_colors) = levels(sobj$seurat_clusters)
  custom_colors[!(names(custom_colors) %in% clusters_oi)] = "gray92"
  
  p1 = Seurat::DimPlot(sobj, reduction = name2D, cols = custom_colors,
                       group.by = "seurat_clusters", label = TRUE) +
    ggplot2::labs(title = "Clusters of interest",
                  subtitle = paste0(clusters_oi, collapse = ", ")) +
    ggplot2::theme(aspect.ratio = 1,
                   plot.title = element_text(hjust = 0.5),
                   plot.subtitle = element_text(hjust = 0.5)) +
    Seurat::NoAxes() + Seurat::NoLegend()
  
  # Color for cell type
  custom_colors = unlist(color_markers)
  custom_colors[!(names(custom_colors) %in% pop_oi)] = "gray92"
  
  p2 = Seurat::DimPlot(sobj, reduction = name2D,
                       group.by = "cell_type", cols = custom_colors) +
    ggplot2::labs(title = "Annotation of interest",
                  subtitle = pop_oi) +
    ggplot2::theme(aspect.ratio = 1,
                   plot.title = element_text(hjust = 0.5),
                   plot.subtitle = element_text(hjust = 0.5)) +
    Seurat::NoAxes() + Seurat::NoLegend()
  
  # Piechart for each cluster, by sample name
  plot_list = lapply(clusters_oi, FUN = function(one_cluster) {
    df = sobj@meta.data %>%
      dplyr::filter(.data$seurat_clusters == .env$one_cluster) %>%
      dplyr::select(sample_identifier, seurat_clusters)
    
    p = aquarius::plot_piechart(df,
                                grouping_var = "sample_identifier",
                                colors = sample_info$color) +
      ggplot2::labs(title = paste0("Cluster ", one_cluster),
                    subtitle = paste0(nrow(df), " cells")) +
      ggplot2::theme(plot.title = element_text(hjust = 0.5),
                     plot.subtitle = element_text(hjust = 0.5))
    
    return(p)
  })
  p3 = patchwork::wrap_plots(plot_list)
  
  # Patchwork
  p = patchwork::wrap_plots(p2, p1, p3, nrow = 1)
  
  return(p)
}
```


# Visualization

## Gene expression

We visualize gene expression for some markers :

```{r plot_list_features, fig.width = 12, fig.height = 4}
features = c("percent.mt", "percent.rb", "nFeature_RNA")

plot_list = lapply(features, FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

## Clusters

We visualize clusters :

```{r see_clustering, fig.width = 6, fig.height = 4}
cluster_plot = Seurat::DimPlot(sobj, reduction = name2D, label = TRUE) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1)
cluster_plot
```

## Cell type

We visualize cell type split by sample :

```{r plot_split_dimred, fig.width = 12, fig.height = 10}
plot_list = aquarius::plot_split_dimred(sobj,
                                        reduction = name2D,
                                        split_by = "project_name",
                                        group_by = "cell_type",
                                        split_color = setNames(sample_info$color,
                                                               nm = sample_info$project_name),
                                        group_color = color_markers,
                                        bg_pt_size = 0.5, main_pt_size = 0.5)

plot_list[[length(plot_list) + 1]] = cluster_plot

patchwork::wrap_plots(plot_list, ncol = 4) &
  Seurat::NoLegend()
```

## Cluster type

We summarize major cell type by cluster :

```{r cell_type_clusters}
sobj$cell_type = as.factor(sobj$cell_type)

cell_type_clusters = sobj@meta.data[, c("cell_type", "seurat_clusters")] %>%
  table() %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cell_type_clusters = setNames(levels(sobj$cell_type)[cell_type_clusters],
                              nm = names(cell_type_clusters))

```

We define cluster type :

```{r table_cluster_type}
sobj$cluster_type = cell_type_clusters[sobj$seurat_clusters] %>%
  as.factor()
table(sobj$cluster_type, sobj$cell_type)
```

We compare cluster annotation and cell type annotation :

```{r see_cluster_type, fig.width = 10, fig.height = 5}
p1 = Seurat::DimPlot(sobj, group.by = "cell_type",
                     reduction = name2D, cols = color_markers) +
  ggplot2::labs(title = "Cell type") +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5))

p2 = Seurat::DimPlot(sobj, group.by = "cluster_type",
                     reduction = name2D, cols = color_markers) +
  ggplot2::labs(title = "Cluster type") +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5))

patchwork::wrap_plots(p1, p2, guides = "collect")
```

# Differential expression

What is the identity of specific clusters ?

## Dendritic cells

DC are those clusters :

```{r clusters_oiDC}
clusters_oi = names(which(cell_type_clusters == "DC"))
clusters_oi
```

We represent those clusters on the projection :

```{r see_DC, fig.width = 12, fig.height = 5}
see_clusters("DC")
```


### Cluster 5 (Ccr7)

```{r cluster5_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 5
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster5_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 8 (Hspa1a/b)

```{r cluster8_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 8
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster8_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 16 (cycling)

```{r cluster16_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 16
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster16_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 19 (Xcr1)

```{r cluster19_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 19
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster19_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 20 (Siglech, Cd7)

```{r cluster20_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 20
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster20_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 22 (Cxcl10/11)

```{r cluster22_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 22
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster22_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 26 (Mreg)

```{r cluster26_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 26
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster26_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


## Macrophages

Macrophages are those clusters :

```{r clusters_oimacrophages}
clusters_oi = names(which(cell_type_clusters == "macrophages"))
clusters_oi
```

We represent those clusters on the projection :

```{r see_macrophages, fig.width = 12, fig.height = 5}
see_clusters("macrophages")
```

### Cluster 10 (Plac8, M1)

```{r cluster10_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 10
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster10_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 11 (Cytip, mono ?)

```{r cluster11_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 11
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster11_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 12 (Cmpk2)

```{r cluster12_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 12
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster12_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 0 (Rsad2 high)

```{r cluster0_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 0
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster0_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Clusters 3 and 7 (Retnla high, epi-macro)

```{r cluster3_7_see_blue_red, fig.width = 5, fig.height = 5}
group1 = c(3, 7)
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster3_7_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 2 (Klf2 high)

```{r cluster2_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 2
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster2_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 6 (mito)

```{r cluster6_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 6
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster6_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 15 (pNF, Cx3xcr1)

```{r cluster15_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 15
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster15_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 1 (Il1b)

```{r cluster1_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 1
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster1_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 4 (Cx3cr1)

```{r cluster4_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 4
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster4_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


## T cells

T cells are those clusters :

```{r clusters_oit}
clusters_oi = names(which(cell_type_clusters == "T cells"))
clusters_oi
```

We represent those clusters on the projection :

```{r see_t, fig.width = 12, fig.height = 5}
see_clusters("T cells")
```

### Cluster 13 (Themis)

```{r cluster13_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 13
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster13_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


### Cluster 14 (Gata3)

```{r cluster14_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 14
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster14_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 23 (Scart1)

```{r cluster23_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 23
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster23_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 24 (Igfbp7)

```{r cluster24_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 24
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster24_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```

### Cluster 25 (Foxp3)

```{r cluster25_see_blue_red, fig.width = 5, fig.height = 5}
group1 = 25
group2 = clusters_oi

aquarius::plot_red_and_blue(sobj, group1, group2, reduction = name2D)
```

We make identify specific markers for these group :

```{r cluster25_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj,
                           ident.1 = group1, ident.2 = group2)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-avg_logFC, pct.1 - pct.2)
dim(mark)

head(mark, n = 20)
```


# Manual annotation

We re-annotate cell types. But, they can more sub-annotated :

* macrophages :
  - M1 macrophages (Plac8, Chil3)
  - M2 epineurial macrophages (Retnla) : Rsad2 high or Klf2 high
  - M2 endoneurial macrophages (Bcl2a1b) : 
      - Il1b population
      - Cx3cr1 population

* monocytes :
  - Cytip population
  - Cmpk2 population

* dendritic cells :
  - Mreg population (https://doi.org/10.7554/eLife.56890, https://www.nature.com/articles/s41586-020-2134-y)
  - mDC (Cd209a+) : Hspa1a population, Ccr7 population and Cxcl10 population
  - pDC (Siglech+)
  - cDC (Xcr1+)

* T cells :
  - Themis+
  - Gata3+
  - Foxp3+
  - Scart1+
  - Igfbp7+

* NK cells :
  - Prf1+ population
  - Prf1- population


```{r cell_type_manual}
sobj$cell_type_manual = sobj@meta.data %>%
  dplyr::mutate(cell_type_manual = case_when(
    # Macrophages
    seurat_clusters %in% c(0, 3, 7, 2, 6, 15) ~ "Epineurial macrophages (M2)",
    seurat_clusters %in% c(1) ~ "Endoneurial macrophages (M2)",
    seurat_clusters %in% c(4) ~ "Endoneurial macrophages (M2) Cx3cr1+",
    seurat_clusters %in% c(10) ~ "M1 macrophages",
    seurat_clusters %in% c(11, 12) ~ "monocytes-macrophages",
    
    # Dendritic cells
    seurat_clusters %in% c(8, 5, 22, 16) ~ "mDC",
    seurat_clusters %in% c(19) ~ "cDC",
    seurat_clusters %in% c(20) ~ "pDC",
    seurat_clusters %in% c(26) ~ "mregDC",
    
    # Rare pop
    seurat_clusters %in% c(21) ~ "B cells",
    seurat_clusters %in% c(17) ~ "mast cells",
    seurat_clusters %in% c(18) ~ "neutrophils",
    seurat_clusters %in% c(9) ~ "NK cells",
    
    # T cells and NK cells
    seurat_clusters %in% c(13) ~ "T cells",
    seurat_clusters %in% c(25) ~ "T reg",
    seurat_clusters %in% c(24, 23, 14) ~ "T eff")) %>%
  dplyr::pull(cell_type_manual) %>%
  as.factor()

table(sobj$cell_type_manual,
      sobj$cell_type)
```

We associate custom colors :

```{r manual_palette}
manual_palette = list("Epineurial macrophages (M2)" = aquarius::c30_palette[3],
                      "Endoneurial macrophages (M2)" = aquarius::c30_palette[22],
                      "Endoneurial macrophages (M2) Cx3cr1+" = aquarius::c30_palette[23],
                      "M1 macrophages" = aquarius::c30_palette[28],
                      "monocytes-macrophages" = aquarius::c30_palette[10],
                      "mDC" = aquarius::c30_palette[5],
                      "cDC" = aquarius::c30_palette[12],
                      "pDC" = aquarius::c30_palette[14],
                      "mregDC" = aquarius::c30_palette[23],
                      "B cells" = color_markers$`B cells`,
                      "mast cells" = color_markers$`mast cells`,
                      "neutrophils" = color_markers$neutrophils,
                      "NK cells" = color_markers$`NK cells`,
                      "T cells" = aquarius::c30_palette[4],
                      "T reg" = aquarius::c30_palette[26],
                      "T eff" = aquarius::c30_palette[27])

sobj$cell_type_manual = as.character(sobj$cell_type_manual) %>%
  factor(., levels = names(manual_palette))
```

```{r cell_type_manual_see, fig.width = 7, fig.height = 7}
p = Seurat::DimPlot(sobj, reduction = name2D,
                    group.by = "cell_type_manual", label = TRUE,
                    cols = manual_palette) +
  ggplot2::labs(title = "Immune cells atlas") +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_text(hjust = 0.5),
                 legend.position = "none")
p
```

We make a representation split by sample :

```{r plot_split_dimred_new_annot, fig.width = 10, fig.height = 15}
plot_list = aquarius::plot_split_dimred(sobj,
                                        reduction = name2D,
                                        split_by = "project_name",
                                        group_by = "cell_type_manual",
                                        split_color = setNames(sample_info$color,
                                                               nm = sample_info$project_name),
                                        group_color = manual_palette)

plot_list[[length(plot_list) + 1]] = patchwork::guide_area()

patchwork::wrap_plots(plot_list, ncol = 3) +
  patchwork::plot_layout(guides = "collect") &
  guides(color = guide_legend(override.aes = list(size = 3), title = "Cluster annotation"))
```


We represent also the information as a barplot :

```{r fig.width = 8, fig.height = 8}
df_proportion = as.data.frame(table(sobj$sample_identifier, sobj$cell_type_manual))
colnames(df_proportion) = c("orig.ident", "groupby", "freq")

aquarius::plot_barplot(df_proportion,
                       x = "orig.ident",
                       y = "freq",
                       fill = "groupby",
                       position = ggplot2::position_fill()) +
  ggplot2::scale_fill_manual(name = "Cell type",
                             breaks = levels(sobj$cell_type_manual),
                             values = manual_palette) +
  ggplot2::theme(legend.position = "bottom")
```

And as a heatmap :

```{r prop_heatmap_manual, fig.width = 12, fig.height = 6}
# Major cell type by cluster
cell_type_clusters = sobj@meta.data[, c("cell_type", "cell_type_manual")] %>%
  table() %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cell_type_clusters = setNames(levels(sobj$cell_type)[cell_type_clusters],
                              nm = names(cell_type_clusters))

# Heatmap annotation
ha = ComplexHeatmap::HeatmapAnnotation(cell_type = cell_type_clusters,
                                       col = list(cell_type = unlist(color_markers)))

# Draw heatmap
my_mat = sobj@meta.data[, c("sample_identifier", "cell_type_manual")] %>%
  table() %>%
  prop.table(., margin = 1) %>% #"+"(0.001) %>% log()
  as.data.frame.array() %>%
  apply(., 2, FUN = aquarius::run_rescale, new_min = 0, new_max = 1)

ht = ComplexHeatmap::Heatmap(my_mat,
                             row_names_gp = grid::gpar(names = sample_info$sample_identifier,
                                                       col = sample_info$color),
                             row_names_side = "left",
                             top_annotation = ha, cluster_rows = FALSE)
ComplexHeatmap::draw(ht, show_heatmap_legend = FALSE)
```

Group by condition ?

```{r prop_heatmap_condition, fig.width = 12, fig.height = 6}
ht = ComplexHeatmap::Heatmap(my_mat,
                             row_names_gp = grid::gpar(names = sample_info$sample_identifier,
                                                       col = sample_info$color),
                             row_names_side = "left",
                             top_annotation = ha, cluster_rows = TRUE)
ComplexHeatmap::draw(ht, show_heatmap_legend = FALSE)
```

# Save

We save the Seurat object :

```{r save_sobj}
saveRDS(sobj, file = paste0(out_dir, "/", save_name, "_sobj_reannot.rds"))
```

and the new colors :

```{r save_color}
saveRDS(manual_palette, file = paste0(out_dir, "/", save_name, "_color_manual.rds"))
```

# R Session

```{r sessioninfo, echo = FALSE, fold_output = TRUE}
sessionInfo()
```

