---
title: "HS project"
subtitle: "Zoom in hair follicle stem cells (HF-SCs)"
author: "Audrey"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot"),
  time_it = local({
    now = NULL
    function(before, options) {
      if (options$time_it) {
        if (before) {
          now <= Sys.time()
        } else {
          res = difftime(Sys.time(), now, units = "secs")
          paste("(Time to run :", round(res, digits = 2), "s)")
        }
      }
    }
  })
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1337L)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # usefull for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15,
                      
                      # something about seed, chunk and Rmarkdown compilation
                      # https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script
                      # cache = TRUE,
                      cache.lazy = FALSE, 
                      
                      # add runtime after chunk
                      time_it = FALSE)
```


This file is used to analyse the immune cells dataset.

```{r library}
library(dplyr)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)

.libPaths()
```

# Preparation

In this section, we set the global settings of the analysis. We will store data there :

```{r out_dir}
save_name = "hfsc"
out_dir = "."
```

We load the dataset :

```{r load_sobj}
sobj = readRDS(paste0(out_dir, "/", save_name, "_sobj.rds"))
sobj
```

We load the sample information :

```{r custom_palette_sample, fig.width = 6, fig.height = 6}
sample_info = readRDS(paste0(out_dir, "/../../1_metadata/hs_hd_sample_info.rds"))
project_names_oi = sample_info$project_name

graphics::pie(rep(1, nrow(sample_info)),
              col = sample_info$color,
              labels = sample_info$project_name)
```

Here are custom colors for each cell type :

```{r color_markers, fig.width = 10, fig.height = 1.2, class.source = "fold-hide"}
color_markers = readRDS(paste0(out_dir, "/../../1_metadata/hs_hd_color_markers.rds"))

data.frame(cell_type = names(color_markers),
           color = unlist(color_markers)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(color_markers), breaks = names(color_markers)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank(),
                 axis.text.x = element_text(angle = 30, hjust = 1))
```

This is the projection of interest :

```{r name2D}
name2D = "harmony_24_tsne"
```

We design a custom function to make the GSEA plot and a word cloud graph :

```{r make_gsea_plot, class.source = "fold-hide"}
make_gsea_plot = function(gsea_results, gs_oi, fold_change, metric = "FC") {
  fold_change$metric = fold_change[, metric]
  
  plot_list = lapply(gs_oi, FUN = function(gene_set) {
    # Gene set content
    gs_content = gene_sets %>%
      dplyr::filter(gs_name == gene_set) %>%
      dplyr::pull(ensembl_gene) %>%
      unique()
    
    # Gene set size
    nb_genes = length(gs_content)
    
    # Enrichment metrics
    NES = gsea_results@result[gene_set, "NES"]
    p.adjust = gsea_results@result[gene_set, "p.adjust"] %>%
      round(., 4)
    qvalues = gsea_results@result[gene_set, "qvalues"]
    
    if (p.adjust > 0.05) {
      p.adjust = paste0("<span style='color:red;'>", p.adjust, "</span>")
    }
    
    my_subtitle = paste0("\nNES : ", round(NES, 2),
                         " | padj : ", p.adjust,
                         " | qval : ", round(qvalues, 4),
                         " | set size : ", nb_genes, " genes")
    
    # Size limits
    lower_FC = min(fold_change[gs_content, ]$metric, na.rm = TRUE)
    upper_FC = max(fold_change[gs_content, ]$metric, na.rm = TRUE)
    
    # Plot
    p = enrichplot::gseaplot2(x = gsea_results, geneSetID = gene_set) +
      ggplot2::labs(title = gene_set,
                    subtitle = my_subtitle) +
      ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                               margin = ggplot2::margin(3, 3, 5, 3)),
                     plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                              size = 10))
    
    wc = ggplot2::ggplot(fold_change[gs_content, ],
                         aes(label = gene_name, size = abs(metric), color = metric)) +
      ggwordcloud::geom_text_wordcloud_area(show.legend = TRUE) +
      ggplot2::scale_color_gradient2(
        name = metric,
        low = aquarius::color_cnv[1],
        mid = "gray70", midpoint = 0,
        high = aquarius::color_cnv[3]) +
      ggplot2::scale_size_area(max_size = 7) +
      ggplot2::theme_minimal() +
      ggplot2::guides(size = "none")
    
    return(list(p, wc))
  }) %>% unlist(., recursive = FALSE)
  
  return(plot_list)
}
```

# Visualization

## Gene expression

We visualize gene expression for some markers :

```{r plot_list_features, fig.width = 12, fig.height = 4}
features = c("percent.mt", "percent.rb", "nFeature_RNA")

plot_list = lapply(features, FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

## Clusters

We visualize clusters :

```{r see_clustering, fig.width = 6, fig.height = 4}
cluster_plot = Seurat::DimPlot(sobj, reduction = name2D, label = TRUE) +
  Seurat::NoAxes() +
  ggplot2::theme(aspect.ratio = 1)
cluster_plot
```

We visualize cluster split by sample :

```{r plot_split_dimred, fig.width = 12, fig.height = 7}
plot_list = aquarius::plot_split_dimred(sobj,
                                        reduction = name2D,
                                        split_by = "project_name",
                                        group_by = "seurat_clusters",
                                        split_color = setNames(sample_info$color,
                                                               nm = sample_info$project_name),
                                        bg_pt_size = 0.5, main_pt_size = 0.5)

plot_list[[length(plot_list) + 1]] = cluster_plot

patchwork::wrap_plots(plot_list, ncol = 4) &
  Seurat::NoLegend()
```

# Analysis

We make a differential expression between each cluster and all others. Then, we compute the differences between HS and HD, all cells considered.

We save the results in a list :

```{r list_results}
list_results = list()
```

For the differences between HS and HD, we make GSEA using gene sets from MSigDB :

```{r gene_sets}
gene_sets = aquarius::get_gene_sets(species = "Homo sapiens")
gene_sets = gene_sets$gene_sets

head(gene_sets)
```

How many gene sets ?

```{r count_gene_sets}
gene_sets[, c("gs_subcat", "gs_name")] %>%
  dplyr::distinct() %>%
  dplyr::pull(gs_subcat) %>%
  table() %>%
  as.data.frame.table() %>%
  `colnames<-`(c("Category", "Nb gene sets"))
```

We get gene name and gene ID correspondence :

```{r gene_corresp}
gene_corresp = sobj@assays[["RNA"]]@meta.features[, c("gene_name", "Ensembl_ID")] %>%
  `colnames<-`(c("NAME", "ID")) %>%
  dplyr::mutate(ID = as.character(ID))
rownames(gene_corresp) = gene_corresp$ID

head(gene_corresp)
```

## Clusters identity

### Cluster 0

We set the cluster of interest :

```{r clusters_oi_clust0}
clusters_oi = 0
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust0, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust0, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::filter(abs(pct.1 - pct.2) > 0.2) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust0_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 1

We set the cluster of interest :

```{r clusters_oi_clust1}
clusters_oi = 1
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust1, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust1, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust1_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```


### Cluster 2

We set the cluster of interest :

```{r clusters_oi_clust2}
clusters_oi = 2
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust2, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust2, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust2_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 3

We set the cluster of interest :

```{r clusters_oi_clust3}
clusters_oi = 3
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust3, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust3, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust3_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 4

We set the cluster of interest :

```{r clusters_oi_clust4}
clusters_oi = 4
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust4, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust4, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust4_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 5

We set the cluster of interest :

```{r clusters_oi_clust5}
clusters_oi = 5
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust5, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust5, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust5_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 6

We set the cluster of interest :

```{r clusters_oi_clust6}
clusters_oi = 6
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust6, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust6, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust6_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 7

We set the cluster of interest :

```{r clusters_oi_clust7}
clusters_oi = 7
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust7, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust7, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust7_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```

### Cluster 8

We set the cluster of interest :

```{r clusters_oi_clust8}
clusters_oi = 8
table(sobj$seurat_clusters, sobj$sample_type)[clusters_oi + 1, ]
```

We visualize these cells on the projection :

```{r see_clust8, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = clusters_oi)
```

What is the cluster identity ?

```{r ident_clust8, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = clusters_oi)

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05) %>%
  dplyr::filter(avg_logFC > 0) %>%
  dplyr::arrange(-(pct.1 - pct.2), -avg_logFC)

list_results[[paste0("cluster_", clusters_oi, "_vs_all")]] = mark

dim(mark)
head(mark, n = 20)
```

We visualize expression for top 9 genes :

```{r see_clust8_top9, fig.width = 12, fig.height = 12}
plot_list = lapply(rownames(mark)[c(1:9)] %>% na.omit(), FUN = function(one_gene) {
  Seurat::FeaturePlot(sobj, features = one_gene,
                      reduction = name2D) +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    Seurat::NoAxes()
})

patchwork::wrap_plots(plot_list, ncol = 3)
```


## HS vs HD

We change the cells identity :

```{r change_ident}
Seurat::Idents(sobj) = sobj$sample_type
table(Seurat::Idents(sobj))
```

We compare all cells between HS and HD :

```{r see_hs_hd, fig.width = 7, fig.height = 7}
aquarius::plot_red_and_blue(sobj, reduction = name2D,
                            group1 = "HS", group2 = "HD")
```

### Differential expression

We make a differential expression analysis between cells from HS and cells from HD :

```{r hs_vs_hd_de, fig.width = 5, fig.height = 5}
mark = Seurat::FindMarkers(sobj, ident.1 = "HS", ident.2 = "HD")

mark = mark %>%
  dplyr::filter(p_val_adj < 0.05)

list_results[["HS_vs_HD"]] = mark
```

How many differentially expressed genes by condition ?

```{r nb_genes}
# Enriched in HS
mark %>%
  dplyr::filter(avg_logFC > 0) %>%
  nrow()

# Enriched in HD
mark %>%
  dplyr::filter(avg_logFC < 0) %>%
  nrow()
```

There are no significantly enriched genes in HD compared to HS. What are the genes up-regulated in HS ?

```{r enriched_hs}
genes = mark %>%
  dplyr::filter(avg_logFC > 0) %>%
  rownames %>%
  sort()

genes
```

### Heatmap

We visualize the expression of these genes on a heatmap.

```{r prep_genes}
features_oi = genes

length(features_oi)
```

We prepare the scaled expression matrix :

```{r mat}
mat_expression = Seurat::GetAssayData(sobj, assay = "RNA", slot = "data")[features_oi, ]
mat_expression = Matrix::t(mat_expression)
mat_expression = dynutils::scale_quantile(mat_expression) # between 0 and 1
mat_expression = Matrix::t(mat_expression)
mat_expression = as.matrix(mat_expression) # not sparse

dim(mat_expression)
```

We prepare the heatmap annotation :

```{r ha_top}
ha_top = ComplexHeatmap::HeatmapAnnotation(
  sample_type = sobj$sample_type,
  cluster = sobj$seurat_clusters,
  col = list(sample_type = setNames(nm = c("HS", "HD"),
                                    c("#C55F40", "#2C78E6")),
             cluster = setNames(nm = levels(sobj$seurat_clusters),
                                aquarius::gg_color_hue(length(levels(sobj$seurat_clusters))))))
```

And the heatmap :

```{r heatmap, fig.width = 15, fig.height = 40}
sobj$cell_group = paste0(sobj$sample_type, "_", sobj$seurat_clusters) %>%
  as.factor()

ht = ComplexHeatmap::Heatmap(mat_expression,
                             col = aquarius::color_cnv,
                             # Annotation
                             top_annotation = ha_top,
                             # Grouping
                             column_order = sobj@meta.data %>%
                               dplyr::arrange(sample_type, seurat_clusters) %>%
                               rownames(),
                             column_split = sobj$cell_group,
                             column_gap = unit.c(rep(unit(0.01, "mm"), 8),
                                                 unit(2, "mm"),
                                                 rep(unit(0.01, "mm"), 8)),
                             column_title = NULL,
                             cluster_rows = TRUE,
                             cluster_columns = FALSE,
                             show_column_names = FALSE,
                             # Visual aspect
                             show_heatmap_legend = TRUE,
                             border = TRUE)

ComplexHeatmap::draw(ht,
                     merge_legend = TRUE,
                     heatmap_legend_side = "bottom",
                     annotation_legend_side = "bottom")
```

### GSEA

We run a GSEA for all gene sets, from the full count matrix :

```{r gsea_hs_vs_hd, fig.width = 12, fig.height = 40}
ranked_gene_list = aquarius::run_foldchange(Seurat::GetAssayData(sobj, assay = "RNA", slot = "counts"),
                                            group1 = colnames(sobj)[sobj@active.ident %in% "HS"],
                                            group2 = colnames(sobj)[sobj@active.ident %in% "HD"])
names(ranked_gene_list) = gene_corresp$ID

gsea_results = aquarius::gsea_run(ranked_gene_list = ranked_gene_list,
                                  gene_sets = gene_sets[, c("gs_name", "ensembl_gene")],
                                  GSEA_p_val_thresh = 1)

list_results[["HS_vs_HD_gsea"]] = gsea_results

gsea_results@result %>%
  dplyr::filter(pvalue < 0.05) %>%
  dplyr::top_n(., n = 200, wt = abs(NES)) %>%
  dplyr::mutate(too_long = ifelse(nchar(ID) > 60, yes = TRUE, no = FALSE)) %>%
  dplyr::mutate(ID = stringr::str_sub(ID, end = 60)) %>%
  dplyr::mutate(ID = ifelse(too_long, yes = paste0(ID, "..."), no = ID)) %>%
  aquarius::gsea_plot(show_legend = TRUE) +
  ggplot2::labs(title = "GSEA using all genes (count matrix)") +
  ggplot2::theme(plot.title = element_text(size = 20))
```

We compute a fold change table to make a wordcloud :

```{r fold_change_ifegs_hs_vs_hd}
fold_change = gene_corresp
colnames(fold_change) = c("gene_name", "ID")
rownames(fold_change) = fold_change$ID
fold_change$FC = ranked_gene_list[rownames(fold_change)]
fold_change$pct.1 = Seurat::GetAssayData(sobj, assay = "RNA", slot = "counts")[, sobj@active.ident == "HS"] %>%
  apply(., MARGIN = 1, FUN = function(one_row) {
    return( mean(one_row != 0) )
  })
fold_change$pct.2 = Seurat::GetAssayData(sobj, assay = "RNA", slot = "counts")[, sobj@active.ident == "HD"] %>%
  apply(., MARGIN = 1, FUN = function(one_row) {
    return( mean(one_row != 0) )
  })
fold_change$FC_x_pct = ifelse(fold_change$FC > 0,
                              yes = fold_change$FC * fold_change$pct.1,
                              no = fold_change$FC * fold_change$pct.2)

dim(fold_change) ; head(fold_change)
```

We make the gsea plot for some gene sets :

```{r gsea_plot_within_ifegs_hs_vs_hd, fig.width = 12, fig.height = 25}
gs_oi = c("GOMF_STRUCTURAL_CONSTITUENT_OF_RIBOSOME",
          "GOBP_B_CELL_CHEMOTAXIS",
          "GOBP_AUTOCRINE_SIGNALING",
          "GOBP_COMPLEMENT_ACTIVATION",
          "GOBP_CORNIFICATION",
          "WP_HAIR_FOLLICLE_DEVELOPMENT_ORGANOGENESIS_PART_2_OF_3",
          "PID_INTEGRIN3_PATHWAY")

plot_list = make_gsea_plot(gsea_results, gs_oi, fold_change, "FC_x_pct")

patchwork::wrap_plots(plot_list, ncol = 2, widths = c(2,1.5))
```

# Save

We save the list of results :

```{r save_list_results}
saveRDS(list_results, file = paste0(out_dir, "/", save_name, "_list_results.rds"))
```


# R Session

```{r sessioninfo, echo = FALSE, fold_output = TRUE}
sessionInfo()
```

