---
title: "HS project"
subtitle: "Figures"
author: "Audrey"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    code_folding: show
    code_download: true
    toc: true
    toc_float: true
    number_sections: false
---

<style>
body {
text-align: justify}
</style>

<!-- Automatically computes and prints in the output the running time for any code chunk -->
```{r, echo=FALSE}
# https://github.com/rstudio/rmarkdown/issues/1453
hooks = knitr::knit_hooks$get()
hook_foldable = function(type) {
  force(type)
  function(x, options) {
    res = hooks[[type]](x, options)
    
    if (isFALSE(options[[paste0("fold_", type)]])) return(res)
    
    paste0(
      "<details><summary>", "show", "</summary>\n\n",
      res,
      "\n\n</details>"
    )
  }
}
knitr::knit_hooks$set(
  output = hook_foldable("output"),
  plot = hook_foldable("plot"),
  time_it = local({
    now = NULL
    function(before, options) {
      if (options$time_it) {
        if (before) {
          now <= Sys.time()
        } else {
          res = difftime(Sys.time(), now, units = "secs")
          paste("(Time to run :", round(res, digits = 2), "s)")
        }
      }
    }
  })
)
```

<!-- Set default parameters for all chunks -->
```{r, setup, include = FALSE}
set.seed(1337L)
knitr::opts_chunk$set(echo = TRUE, # display code
                      # display chunk output
                      message = FALSE,
                      warning = FALSE,
                      fold_output = FALSE, # useful for sessionInfo()
                      fold_plot = FALSE,
                      
                      # figure settings
                      fig.align = 'center',
                      fig.width = 20,
                      fig.height = 15,
                      
                      # something about seed, chunk and Rmarkdown compilation
                      # https://stackoverflow.com/questions/39417003/long-vectors-not-supported-yet-error-in-rmd-but-not-in-r-script
                      # cache = TRUE,
                      cache.lazy = FALSE, 
                      
                      # add runtime after chunk
                      time_it = FALSE,
                      
                      # save figures in PDF in a separate folder
                      dev = c('png', 'pdf'), # tiff or pdf alone renders bad in html
                      # dpi = 300,
                      fig.path = "figures_detail/",
                      pdf.options(encoding = "ISOLatin9.enc"))
```


This file is used to prepare the figures for the paper.

```{r library}
library(dplyr)
library(patchwork)
library(ggplot2)
library(ComplexHeatmap)

.libPaths()
```


# Preparation

Here are the folders where analyzes are stored :

```{r locations}
data_dir = "./.."
list.files(data_dir)
```

These are the custom colors for cell populations :

```{r color_markers, fig.width = 10, fig.height = 1.2, class.source = "fold-hide"}
color_markers = readRDS(paste0(data_dir, "/1_metadata/hs_hd_color_markers.rds"))
color_markers = color_markers[names(color_markers) != "melanocytes"] # remove melanocytes

data.frame(cell_type = names(color_markers),
           color = unlist(color_markers)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(color_markers), breaks = names(color_markers)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank(),
                 axis.text.x = element_text(angle = 30, hjust = 1))
```

We define custom colors for sample type :

```{r sample_type_colors, fig.width = 3, fig.height = 0.75, class.source = "fold-hide"}
sample_type_colors = setNames(nm = c("HS", "HD"),
                              c("#C55F40", "#2C78E6"))

data.frame(cell_type = names(sample_type_colors),
           color = unlist(sample_type_colors)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(sample_type_colors), breaks = names(sample_type_colors)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank())
```

We define custom colors for laboratory :

```{r laboratory_colors, fig.width = 3, fig.height = 0.75, class.source = "fold-hide"}
laboratory_colors = setNames(nm = c("Our", "Wu", "Takahashi"),
                             c("firebrick3", "deepskyblue", "mediumpurple"))

data.frame(cell_type = names(laboratory_colors),
           color = unlist(laboratory_colors)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(laboratory_colors), breaks = names(laboratory_colors)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank())
```

We define custom colors for location :

```{r location_colors, fig.width = 3, fig.height = 0.75, class.source = "fold-hide"}
location_colors = setNames(nm = c("axillary", "hair scalp", "pubis"),
                           c("forestgreen", "orchid3", "darkorange1"))

data.frame(cell_type = names(location_colors),
           color = unlist(location_colors)) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0, fill = cell_type)) +
  ggplot2::geom_point(pch = 21, size = 5) +
  ggplot2::scale_fill_manual(values = unlist(location_colors), breaks = names(location_colors)) +
  ggplot2::theme_classic() +
  ggplot2::theme(legend.position = "none",
                 axis.line = element_blank(),
                 axis.title = element_blank(),
                 axis.ticks = element_blank(),
                 axis.text.y = element_blank())
```

We set a background color :

```{r bg_color}
bg_color = "gray94"
```


This is the correspondence between cell types and cell families, and custom colors to color cells by cell category :

```{r cell_category}
custom_order_cell_type = data.frame(
  cell_type = names(color_markers),
  cell_category = c(rep("immune cells", 5),
                    rep("matrix", 5),
                    rep("non matrix", 5)),
  stringsAsFactors = FALSE)
custom_order_cell_type$cell_type = factor(custom_order_cell_type$cell_type,
                                          levels = custom_order_cell_type$cell_type)
rownames(custom_order_cell_type) = custom_order_cell_type$cell_type
custom_order_cell_type

category_color = c("immune cells" = "slateblue1",
                   "matrix" = "mediumseagreen",
                   "non matrix" = "firebrick3")
```

We define markers to display on a dotplot to assess cell type annotation :

```{r dotplot_markers}
dotplot_markers = c("PTPRC",                # immune cells
                    "CD3E", "CD4",          # CD4+ T cells
                    "CD3E", "CD8A",         # CD8+ T cells
                    "CD207", "AIF1",        # Langerhans cells
                    "TREM2", "MSR1",        # macrophages
                    "CD79A", "CD79B",       # B cells
                    "MSX2",                 # matrix cells
                    "KRT32", "KRT35",       # cuticle
                    "KRT31", "PRR9",        # cortex
                    "BAMBI", "ALDH1A3",     # medulla
                    "KRT71", "KRT73",       # IRS
                    "TOP2A", "MCM5", "TK1", # cycling cells
                    "KRT14", "CXCL14",      # non-matrix cells
                    "DIO2", "TCEAL2",       # HF-SCs
                    "KRT15", "COL17A1",     # IFE basal
                    "SPINK5", "KRT1",       # ORS
                    "KRT16", "KRT6C",       # IFE granular or spinous
                    "CLMP", "PPARG")        # sebocytes
```

We copy-paste a subset of the markers associated with cell type annotation :

```{r cell_markers}
cell_markers = list(
  #  Immune cells
  "CD4 T cells" = c("CD3D", "CD3G", "CD3E", "CD52", "IL32",
                    "TRBC1", "CD4", "COTL1", "CD40LG", "GPR183", "TNFRSF1B", "FYB1", "LAT"),
  "CD8 T cells" = c("CD3D", "CD3G", "CD3E", "CD52", "IL32",
                    "CTSW", "NCR3", "CD8A", "NKG7", "KLRC1", "TRGC2", "CCL5", "ZNF683"),
  "Langerhans cells" = c("CD207", "LST1", "AIF1", "CPVL", "C15orf48", "CD1A", "CD52", "CD1E", "CD1C"),
  "macrophages" = c("LST1", "AIF1", "C3", "OLFML3", "TREM2", "A2M",
                    "MSR1", "FCGR3A", "VSIG4", "AP003481.1"),
  "B cells" = c("CD79A", "IGHG2", "IGHG3", "IGLC3", "IGHG1", "IGHM", "MS4A1", "IGLC2",
                "IGHGP", "BANK1", "IGHG4", "IGKC", "TNFRSF13C", "CD79B", "KLF2", "GZMB"),
  # Matrix
  "cuticle" = c("KRT32", "KIF21A", "CCND2", "VCAN", "CYSTM1", "TESC", "MT4", "KRT35",
                "DIAPH3", "VSNL1", "NOTCH1", "SNCAIP", "CADM1", "OCA2", "LAMP5"),
  "cortex" = c("KRTAP11-1", "PRR9", "C10orf99", "CRYAB", "HEPHL1", "ALOX15B", "EFHD1", "KRT35",
               "CALHM4", "SPECC1", "DSG4", "DNASE1L2", "TGM3", "KRT31", "KRT36", "KIAA1211L"),
  "medulla" = c("BAMBI", "SLC7A8", "EDNRA", "ALDH1A3", "IGFBP2",
                "KRT25", "KITLG", "MSX2", "SLC6A15", "PRDM1"),
  "IRS" = c("KRT71", "KRT27", "KRT28", "KRT25", "CTSC", "GATA3", "KRT72", "KRT73", "KRT74",
            "SCEL", "GFRA3", "POF1B", "RDH12", "LMO7", "TCHH", "FABP9"),
  "proliferative" = c("NUSAP1", "PCLAF", "TOP2A", "MKI67", "BIRC5", "CENPF",
                      "TK1", "TYMS", "MCM5", "GINS2", "MCM5", "PCLAF",
                      "CDCA7", "MCM3", "TYMS", "CLSPN", "FEN1", "FAM111B", "CDT1"),
  # Non-matrix
  "HF-SCs" = c("KRT15", "DST", "DIO2", "TCEAL2", "SFRP1", "TNC", "WIF1", "FRZB", "LHX2", "SPRY1",
               "COL1A2", "ITGA6", "SOX9", "S100A6", "ANGPTL7", "LGR5", "COL17A1"),
  "IFE basal" = c("GPX2", "IMPA2", "KRT15", "DST", "COL17A1", "MOXD1", "C19orf48", "CHD7",
                  "ALDH3A1", "AHNAK2", "CDH13", "LAMB3", "CAVIN1", "EHF", "CEBPD", "TP53AIP1"),
  "IFE granular spinous" = c("GPX2", "IMPA2", "DEFB1", "LY6D", "EHF", "PDZK1IP1", "S100A7", "FGFBP1", "LYPD3",
                             "CRABP2", "SPINK5", "KRT1", "KRTDAP", "CIDEA", "SUSD4", "FAM83A", "TMEM45A"),
  "ORS" = c("KRT16", "KRT6B", "KRT6C", "NDUFA4L2", "TM4SF1", "APOC1", "SLC1A6",
            "CLCA2", "RHOV", "CBLN2", "HES4", "SMCO4", "MGST1", "LYPD3", "MGP"),
  # other
  "melanocytes" = c("DCT", "KIT", "MLANA", "GPM6B", "EDNRB", "PMEL", "IGFBP7", "MITF", "ZEB2", "APOD"),
  "sebocytes" = c("CLMP", "GLDC", "PPARG", "ACSBG1", "MGST1", "SAA1", "AR", "APMAP"))

lengths(cell_markers)
```

Custom functions to display gene expression on the heatmap :

```{r color_fun, class.source = "fold-hide"}
color_fun = function(one_gene) {
  gene_range = range(ht_annot[, one_gene])
  gene_palette = circlize::colorRamp2(colors = c("#FFFFFF", aquarius::color_gene[-1]),
                                      breaks = seq(from = gene_range[1], to = gene_range[2],
                                                   length.out = length(aquarius::color_gene)))
  return(gene_palette)
}
```

# Load datasets

In a list, we load:

* Seurat object
* name of the 2D projection to visualize cells on
* various third element such as: sample information, trajectory, results from gene analysis...

```{r list_data}
list_data = list()

## Our dataset
# Atlas of all cells
list_data$our_atlas$sobj = readRDS(paste0(data_dir, "/3_combined/hs_hd_sobj.rds"))
list_data$our_atlas$name2D = "harmony_38_tsne"
list_data$our_atlas$sample_info = readRDS(paste0(data_dir, "/1_metadata/hs_hd_sample_info.rds"))
list_data$our_atlas$sobj

# ORS + IFEb dataset
list_data$ors_ifeb$sobj = readRDS(paste0(data_dir, "/4_zoom/3_zoom_ors_ifeb/ors_ifeb_sobj.rds"))
list_data$ors_ifeb$name2D = "harmony_20_tsne"
list_data$ors_ifeb$list_results = readRDS(paste0(data_dir, "/4_zoom/3_zoom_ors_ifeb/ors_ifeb_list_results.rds"))
list_data$ors_ifeb$sobj

# HF-SCs dataset
list_data$hfsc$sobj = readRDS(paste0(data_dir, "/4_zoom/2_zoom_hfsc/hfsc_sobj.rds"))
list_data$hfsc$name2D = "harmony_24_tsne"
list_data$hfsc$list_results = readRDS(paste0(data_dir, "/4_zoom/2_zoom_hfsc/hfsc_list_results.rds"))
list_data$hfsc$sobj

# Non matrix cells dataset
list_data$non_matrix$sobj = readRDS(paste0(data_dir, "/4_zoom/5_zoom_non_matrix/non_matrix_sobj_traj_tinga.rds"))
list_data$non_matrix$name2D = "harmony_dm"
list_data$non_matrix$my_traj = readRDS(paste0(data_dir, "/4_zoom/5_zoom_non_matrix/non_matrix_my_traj_tinga.rds"))
list_data$non_matrix$sobj

# Matrix cells dataset (not considered in the manuscript)
# list_data$matrix$sobj = readRDS(paste0(data_dir, "/4_zoom/6_zoom_matrix/matrix_sobj.rds"))
# list_data$matrix$name2D = "harmony_19_tsne"
# list_data$matrix$list_results = readRDS(paste0(data_dir, "/4_zoom/6_zoom_matrix/matrix_list_results.rds"))
# list_data$matrix$sobj

# Immune cells
list_data$immune$sobj = readRDS(paste0(data_dir, "/4_zoom/1_zoom_immune/immune_cells_sobj.rds"))
list_data$immune$name2D = "harmony_20_tsne"
list_data$immune$list_results = readRDS(paste0(data_dir, "/4_zoom/1_zoom_immune/immune_cells_list_results.rds"))
list_data$immune$sobj

## Wu dataset (OEP002321)
# Atlas of all cells
list_data$wu_atlas$sobj = readRDS(paste0(data_dir, "/5_wu/3_combined/wu_sobj.rds"))
list_data$wu_atlas$name2D = "harmony_39_tsne"
list_data$wu_atlas$sample_info = readRDS(paste0(data_dir, "/5_wu/1_metadata/wu_sample_info.rds"))

# ORS + IFEb dataset
list_data$wu_ors_ifeb$sobj = readRDS(paste0(data_dir, "/5_wu/4_ors_ifeb/ors_ifeb_sobj.rds"))
list_data$wu_ors_ifeb$name2D = "harmony_20_tsne"
list_data$wu_ors_ifeb$list_results = readRDS(paste0(data_dir, "/5_wu/4_ors_ifeb/ors_ifeb_list_results.rds"))
list_data$wu_ors_ifeb$sobj

## Takahashi dataset (GSE129611)
# Atlas of all cells
list_data$takahashi_atlas$sobj = readRDS(paste0(data_dir, "/6_takahashi/3_combined/takahashi_sobj.rds"))
list_data$takahashi_atlas$name2D = "harmony_41_tsne"
list_data$takahashi_atlas$sample_info = readRDS(paste0(data_dir, "/6_takahashi/1_metadata/takahashi_sample_info.rds"))
list_data$takahashi_atlas$sobj

# ORS + IFEb dataset
list_data$takahashi_ors_ifeb$sobj = readRDS(paste0(data_dir, "/6_takahashi/4_ors_ifeb/ors_ifeb_sobj.rds"))
list_data$takahashi_ors_ifeb$name2D = "harmony_23_tsne"
list_data$takahashi_ors_ifeb$list_results = readRDS(paste0(data_dir, "/6_takahashi/4_ors_ifeb/ors_ifeb_list_results.rds"))
list_data$takahashi_ors_ifeb$sobj

## Three datasets combined (Our + OEP002321 + GSE129611)
list_data$combined$sobj = readRDS(paste0(data_dir, "/3_combined/data3_sobj.rds"))
list_data$combined$name2D = "harmony_37_tsne"
list_data$combined$sobj

## Print
str(list_data, max.level = 2)
```


# Figures

## Sample information

Our dataset:

```{r sample_info_our, fig.width = 12, fig.height = 3}
sobj = list_data$our_atlas$sobj
sample_info = list_data$our_atlas$sample_info

# Nb cells by dataset
to_plot = table(sobj$sample_identifier) %>%
  as.data.frame.table(., stringsAsFactors = FALSE) %>%
  `colnames<-`(c("sample_identifier", "nb_cells")) %>%
  dplyr::left_join(x = ., y = sample_info, by = "sample_identifier") 

# patchwork
plot_list = aquarius::fig_plot_gb(to_plot, title = "Available datasets")
patchwork::wrap_plots(plot_list) +
  patchwork::plot_layout(design = "A\nB", heights = c(0.1,5)) &
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 15))
```

Wu dataset:

```{r sample_info_wu, fig.width = 12, fig.height = 3}
sobj = list_data$wu_atlas$sobj
sample_info = list_data$wu_atlas$sample_info

# Nb cells by dataset
to_plot = table(sobj$sample_identifier) %>%
  as.data.frame.table(., stringsAsFactors = FALSE) %>%
  `colnames<-`(c("sample_identifier", "nb_cells")) %>%
  dplyr::left_join(x = ., y = sample_info, by = "sample_identifier") 

# patchwork
plot_list = aquarius::fig_plot_gb(to_plot, title = "Available datasets")
patchwork::wrap_plots(plot_list) +
  patchwork::plot_layout(design = "A\nB", heights = c(0.1,5)) &
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 15))
```

Takahashi dataset:

```{r sample_info_takahashi, fig.width = 12, fig.height = 3}
sobj = list_data$takahashi_atlas$sobj
sample_info = list_data$takahashi_atlas$sample_info

# Nb cells by dataset
to_plot = table(sobj$sample_identifier) %>%
  as.data.frame.table(., stringsAsFactors = FALSE) %>%
  `colnames<-`(c("sample_identifier", "nb_cells")) %>%
  dplyr::left_join(x = ., y = sample_info, by = "sample_identifier") 

# patchwork
plot_list = aquarius::fig_plot_gb(to_plot, title = "Available datasets")
patchwork::wrap_plots(plot_list) +
  patchwork::plot_layout(design = "A\nB", heights = c(0.1,5)) &
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 15))
```

## Annotation smoothing

For each dataset, we smooth the single-cell level cell type annotation at a cluster level.

* Our dataset:

```{r smoothing_atlas_our}
# Select dataset
sobj = list_data$our_atlas$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$our_atlas$sobj = sobj
```

```{r smoothing_non_matrix_our}
# Select dataset
sobj = list_data$non_matrix$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$non_matrix$sobj = sobj
```

```{r smoothing_immune_our}
# Select dataset
sobj = list_data$immune$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$immune$sobj = sobj
```

* Wu dataset:

```{r smoothing_wu}
# Select dataset
sobj = list_data$wu_atlas$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$wu_atlas$sobj = sobj
```

* Takahashi dataset:

```{r smoothing_takahashi}
# Select dataset
sobj = list_data$takahashi_atlas$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$takahashi_atlas$sobj = sobj
```

* Combined dataset:

```{r smoothing_combined}
# Select dataset
sobj = list_data$combined$sobj

# Annotation smoothing
sobj$cell_type = sobj$cell_type %>%
  as.character() %>%
  factor(., levels = names(color_markers))

cluster_type = table(sobj$cell_type, sobj$seurat_clusters) %>%
  prop.table(., margin = 2) %>%
  apply(., 2, which.max)
cluster_type = setNames(nm = names(cluster_type),
                        levels(sobj$cell_type)[cluster_type])

sobj$cluster_type = cluster_type[sobj$seurat_clusters]
sobj$cluster_type = factor(sobj$cluster_type,
                           levels = levels(sobj$cell_type))

# Consider change
list_data$combined$sobj = sobj
```

For the combined dataset, we also define the cell category:

```{r cell_category_combined}
# Select dataset
sobj = list_data$combined$sobj

# Define cell category
sobj$cell_category = custom_order_cell_type[sobj$cell_type, "cell_category"]
table(sobj$cell_type, sobj$cell_category)

# Consider change
list_data$combined$sobj = sobj
```


## Atlas

### Our dataset

We select the dataset:

```{r our_dataset}
sobj = list_data$our_atlas$sobj
name2D = list_data$our_atlas$name2D
sample_info = list_data$our_atlas$sample_info
```

* sample identifier

```{r sample_id_our, fig.width = 4, fig.height = 4}
# Random order
set.seed(1234)
rnd_order = sample(colnames(sobj), replace = FALSE, size = ncol(sobj))

# Extract coordinates
cells_coord = sobj@reductions[[name2D]]@cell.embeddings %>%
  as.data.frame() %>%
  `colnames<-`(c("Dim1", "Dim2"))
cells_coord$sample_identifier = sobj$sample_identifier
cells_coord$location = sobj$location
cells_coord = cells_coord[(rnd_order), ]

# Plot
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = sample_identifier)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = sample_info$color,
                              breaks = sample_info$sample_identifier) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* location

```{r location_our, fig.width = 4, fig.height = 4}
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = location)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = location_colors,
                              breaks = names(location_colors)) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* cell type annotation

```{r cell_type_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* dotplot for cell type annotation

```{r dotplot_our, fig.width = 8, fig.height = 8.5}
plot_list = aquarius::plot_dotplot(sobj,
                                   markers = dotplot_markers,
                                   assay = "RNA", column_name = "cell_type", nb_hline = 0) +
  ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
  ggplot2::theme(legend.position = "left",
                 legend.justification = "bottom",
                 legend.box = "vertical",
                 legend.box.margin = margin(0,70,0,0),
                 axis.title = element_blank(),
                 axis.ticks.x = element_blank(),
                 axis.text.x = element_blank(),
                 axis.line.x = element_blank(),
                 plot.margin = unit(rep(0, 4), "cm"))

p = data.frame(cell_type = factor(levels(sobj$cell_type),
                                  levels = levels(sobj$cell_type))) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0)) +
  ggplot2::geom_point(size = 0) +
  ggplot2::geom_segment(aes(x = 0.5, xend = 5.5, y = 0, yend = 0), size = 6, col = category_color["immune cells"]) +
  ggplot2::geom_segment(aes(x = 5.5, xend = 10.5, y = 0, yend = 0), size = 6, col = category_color["matrix"]) +
  ggplot2::geom_segment(aes(x = 10.5, xend = 15.5, y = 0, yend = 0), size = 6, col = category_color["non matrix"]) +
  ggplot2::scale_y_continuous(expand = c(0,0), limits = c(0,0)) +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.y = element_blank(),
                 axis.ticks.y = element_blank(),
                 axis.title = element_blank(),
                 axis.line.y = element_blank(),
                 axis.text.x = element_text(angle = 45, hjust = 1, size = 10, color = "black"),
                 plot.margin = unit(c(0,0.5,0.5,0), "cm"))

plot_list = patchwork::wrap_plots(plot_list, p,
                                  ncol = 1, heights = c(25, 1))
plot_list
```

* barplot with proportion of cells for 4 cell type, by sample ID

```{r barplot_atlas, fig.width = 9, fig.height = 5}
# Cells of interest
cells_oi = c("ORS", "IFE basal", "CD4 T cells", "macrophages")

# Proportion of cells by sample ID
df_proportion = table(sobj$cluster_type,
                      sobj$sample_identifier) %>%
  prop.table(., margin = 2) %>%
  as.data.frame.table() %>%
  `colnames<-`(c("cluster_type", "sample_identifier", "freq")) %>%
  dplyr::filter(cluster_type %in% cells_oi) %>%
  dplyr::mutate(cluster_type = factor(cluster_type, levels = cells_oi)) %>%
  dplyr::mutate(freq = 100*freq) # percentage

# Plot
ggplot2::ggplot(df_proportion, aes(x = sample_identifier,
                                   y = freq,
                                   fill = cluster_type)) +
  ggplot2::geom_bar(stat = "identity", position = ggplot2::position_dodge(),
                    color = "black", width = 0.7) +
  ggplot2::labs(y = "% of cells by sample") +
  ggplot2::scale_y_continuous(expand = c(0, 0)) +
  ggplot2::scale_fill_manual(values = color_markers[cells_oi],
                             breaks = cells_oi,
                             name = "Cell Type") +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.title.x = element_blank(),
                 axis.line.x = element_line(color = "gray50"),
                 panel.grid.major.y = element_line(),
                 panel.grid.minor.y = element_line(),
                 text = element_text(size = 15))
```

### Wu dataset

We select the dataset:

```{r wu_dataset}
sobj = list_data$wu_atlas$sobj
name2D = list_data$wu_atlas$name2D
sample_info = list_data$wu_atlas$sample_info
```

* sample identifier

```{r sample_id_wu, fig.width = 4, fig.height = 4}
# Random order
set.seed(1234)
rnd_order = sample(colnames(sobj), replace = FALSE, size = ncol(sobj))

# Extract coordinates
cells_coord = sobj@reductions[[name2D]]@cell.embeddings %>%
  as.data.frame() %>%
  `colnames<-`(c("Dim1", "Dim2"))
cells_coord$sample_identifier = sobj$sample_identifier
cells_coord = cells_coord[(rnd_order), ]

# Plot
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = sample_identifier)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = sample_info$color,
                              breaks = sample_info$sample_identifier) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* cell type annotation

```{r cell_type_wu, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_wu, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* dotplot for cell type annotation

```{r dotplot_wu, fig.width = 8, fig.height = 8.5}
plot_list = aquarius::plot_dotplot(sobj,
                                   markers = dotplot_markers,
                                   assay = "RNA", column_name = "cell_type", nb_hline = 0) +
  ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
  ggplot2::theme(legend.position = "left",
                 legend.justification = "bottom",
                 legend.box = "vertical",
                 legend.box.margin = margin(0,70,0,0),
                 axis.title = element_blank(),
                 axis.ticks.x = element_blank(),
                 axis.text.x = element_blank(),
                 axis.line.x = element_blank(),
                 plot.margin = unit(rep(0, 4), "cm"))

p = data.frame(cell_type = factor(levels(sobj$cell_type),
                                  levels = levels(sobj$cell_type))) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0)) +
  ggplot2::geom_point(size = 0) +
  ggplot2::geom_segment(aes(x = 0.5, xend = 5.5, y = 0, yend = 0), size = 6, col = category_color["immune cells"]) +
  ggplot2::geom_segment(aes(x = 5.5, xend = 10.5, y = 0, yend = 0), size = 6, col = category_color["matrix"]) +
  ggplot2::geom_segment(aes(x = 10.5, xend = 15.5, y = 0, yend = 0), size = 6, col = category_color["non matrix"]) +
  ggplot2::scale_y_continuous(expand = c(0,0), limits = c(0,0)) +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.y = element_blank(),
                 axis.ticks.y = element_blank(),
                 axis.title = element_blank(),
                 axis.line.y = element_blank(),
                 axis.text.x = element_text(angle = 45, hjust = 1, size = 10, color = "black"),
                 plot.margin = unit(c(0,0.5,0.5,0), "cm"))

plot_list = patchwork::wrap_plots(plot_list, p,
                                  ncol = 1, heights = c(25, 1))
plot_list
```

### Takahashi dataset

We select the dataset:

```{r takahashi_dataset}
sobj = list_data$takahashi_atlas$sobj
name2D = list_data$takahashi_atlas$name2D
sample_info = list_data$takahashi_atlas$sample_info
```

* sample identifier

```{r sample_id_takahashi, fig.width = 4, fig.height = 4}
# Random order
set.seed(1234)
rnd_order = sample(colnames(sobj), replace = FALSE, size = ncol(sobj))

# Extract coordinates
cells_coord = sobj@reductions[[name2D]]@cell.embeddings %>%
  as.data.frame() %>%
  `colnames<-`(c("Dim1", "Dim2"))
cells_coord$sample_identifier = sobj$sample_identifier
cells_coord = cells_coord[(rnd_order), ]

# Plot
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = sample_identifier)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = sample_info$color,
                              breaks = sample_info$sample_identifier) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* cell type annotation

```{r cell_type_takahashi, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_takahashi, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* dotplot for cell type annotation

```{r dotplot_takahashi, fig.width = 8, fig.height = 8.5}
plot_list = aquarius::plot_dotplot(sobj,
                                   markers = dotplot_markers,
                                   assay = "RNA", column_name = "cell_type", nb_hline = 0) +
  ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
  ggplot2::theme(legend.position = "left",
                 legend.justification = "bottom",
                 legend.box = "vertical",
                 legend.box.margin = margin(0,70,0,0),
                 axis.title = element_blank(),
                 axis.ticks.x = element_blank(),
                 axis.text.x = element_blank(),
                 axis.line.x = element_blank(),
                 plot.margin = unit(rep(0, 4), "cm"))

p = data.frame(cell_type = factor(levels(sobj$cell_type),
                                  levels = levels(sobj$cell_type))) %>%
  ggplot2::ggplot(., aes(x = cell_type, y = 0)) +
  ggplot2::geom_point(size = 0) +
  ggplot2::geom_segment(aes(x = 0.5, xend = 5.5, y = 0, yend = 0), size = 6, col = category_color["immune cells"]) +
  ggplot2::geom_segment(aes(x = 5.5, xend = 10.5, y = 0, yend = 0), size = 6, col = category_color["matrix"]) +
  ggplot2::geom_segment(aes(x = 10.5, xend = 15.5, y = 0, yend = 0), size = 6, col = category_color["non matrix"]) +
  ggplot2::scale_y_continuous(expand = c(0,0), limits = c(0,0)) +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.y = element_blank(),
                 axis.ticks.y = element_blank(),
                 axis.title = element_blank(),
                 axis.line.y = element_blank(),
                 axis.text.x = element_text(angle = 45, hjust = 1, size = 10, color = "black"),
                 plot.margin = unit(c(0,0.5,0.5,0), "cm"))

plot_list = patchwork::wrap_plots(plot_list, p,
                                  ncol = 1, heights = c(25, 1))
plot_list
```

### Combined datasets

We select the dataset:

```{r combined_dataset}
sobj = list_data$combined$sobj
name2D = list_data$combined$name2D
sample_info = rbind(list_data$our_atlas$sample_info,
                    list_data$wu_atlas$sample_info,
                    list_data$takahashi_atlas$sample_info)
```

* sample identifier

```{r sample_id_combined, fig.width = 4, fig.height = 4}
# Random order
set.seed(1234)
rnd_order = sample(colnames(sobj), replace = FALSE, size = ncol(sobj))

# Extract coordinates
cells_coord = sobj@reductions[[name2D]]@cell.embeddings %>%
  as.data.frame() %>%
  `colnames<-`(c("Dim1", "Dim2"))
cells_coord$sample_identifier = sobj$sample_identifier
cells_coord$location = sobj$location
cells_coord$laboratory = sobj$laboratory
cells_coord = cells_coord[(rnd_order), ]

# Plot
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = sample_identifier)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = sample_info$color,
                              breaks = sample_info$sample_identifier) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* location

```{r location_combined, fig.width = 4, fig.height = 4}
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = location)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = location_colors,
                              breaks = names(location_colors)) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* laboratory

```{r laboratory_combined, fig.width = 4, fig.height = 4}
ggplot2::ggplot(cells_coord, aes(x = Dim1, y = Dim2, col = laboratory)) +
  ggplot2::geom_point(size = 0.5) +
  ggplot2::scale_color_manual(values = laboratory_colors,
                              breaks = names(laboratory_colors)) +
  ggplot2::theme_void() +
  ggplot2::theme(aspect.ratio = 1,
                 legend.position = "none")
```

* cell type annotation

```{r cell_type_combined, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_combined, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* split by laboratory:

```{r combined_split, fig.width = 12, fig.height = 5}
plot_list = aquarius::plot_split_dimred(sobj,
                                        reduction = name2D,
                                        split_by = "laboratory",
                                        group_by = "cluster_type",
                                        group_color = color_markers,
                                        order = TRUE,
                                        bg_color = bg_color)

patchwork::wrap_plots(plot_list, nrow = 1) +
  patchwork::plot_layout(guides = "collect") &
  Seurat::NoLegend()
```

* gene to assess global annotation

```{r gene_global_combined, fig.width = 4, fig.height = 4}
plot_list = lapply(c("PTPRC", "MSX2", "KRT14"), FUN = function(one_gene) {
  p = Seurat::FeaturePlot(sobj, features = one_gene, reduction = name2D) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    ggplot2::theme(aspect.ratio = 1) +
    Seurat::NoAxes() +
    Seurat::NoLegend()
  
  return(p)
})

plot_list
```

* violin plot, split by cell type, split by laboratory

for non matrix cells

```{r violin_non_matrix_combined, fig.width = 4, fig.height = 4}
category_of_interest = "non matrix"

# Select cells of interest
cell_type_of_interest = as.character(custom_order_cell_type[
  which(custom_order_cell_type$cell_category == category_of_interest), "cell_type"])

subsobj_of_interest = subset(sobj, cell_type %in% cell_type_of_interest)
subsobj_of_interest$cell_type_labo = paste0(subsobj_of_interest$cell_type,
                                            subsobj_of_interest$laboratory)

cell_markers_of_interest = cell_markers[cell_type_of_interest]

total_height = max(lengths(cell_markers_of_interest)) + 1

# Violin plot
patchwork_list = lapply(names(cell_markers_of_interest), FUN = function(cell_type) {
  markers_set = cell_markers_of_interest[[cell_type]]
  
  # Individual violin plot
  sub_plot_list = lapply(markers_set, FUN = function(one_gene) {
    p = Seurat::VlnPlot(subsobj_of_interest,
                        features = one_gene, pt.size = 0,
                        group.by = "cell_type",
                        split.by = "laboratory",
                        cols = laboratory_colors,
                        combine = FALSE)[[1]] +
      ggplot2::labs(y = one_gene) +
      ggplot2::scale_y_continuous(expand = c(0.01, 0)) +
      ggplot2::geom_vline(mapping = NULL, data = NULL,
                          xintercept = c(1:4) + 0.5,
                          col = "gray50", lty = 2) +
      Seurat::NoLegend() +
      ggplot2::theme(plot.title = element_blank(),
                     axis.title.x = element_blank(), # "Identity"
                     axis.line.x = element_blank(),  # replaced by panel.border
                     axis.title.y = element_text(size = 8, angle = 0, vjust = 0.5), # gene
                     axis.text.y = element_blank(), # "Expression level"
                     axis.text.x = element_blank(), # cell type
                     axis.ticks = element_blank(),
                     plot.margin = unit(c(0,0,0,0), "cm"),
                     panel.border = element_rect(size = 1, color = "gray50"))
    return(p)
  })
  
  # Reset axis.text for last plot
  # sub_plot_list[[length(sub_plot_list)]] = sub_plot_list[[length(sub_plot_list)]] +
  #   ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) # cell type
  
  # Add empty plot to arrange heights
  sub_plot_list[[length(sub_plot_list) + 1]] = patchwork::plot_spacer()
  empty_plot_height = total_height - length(markers_set)
  
  # Arrange as patchwork
  pp = patchwork::wrap_plots(sub_plot_list,
                             ncol = 1,
                             heights = c(rep(1, length(markers_set)),
                                         empty_plot_height)) +
    patchwork::plot_annotation(title = cell_type)
  
  return(pp)
})
names(patchwork_list) = names(cell_markers_of_interest)

patchwork_list
```

for matrix cells

```{r violin_matrix_combined, fig.width = 4, fig.height = 4}
category_of_interest = "matrix"

# Select cells of interest
cell_type_of_interest = as.character(custom_order_cell_type[
  which(custom_order_cell_type$cell_category == category_of_interest), "cell_type"])

subsobj_of_interest = subset(sobj, cell_type %in% cell_type_of_interest)
subsobj_of_interest$cell_type_labo = paste0(subsobj_of_interest$cell_type,
                                            subsobj_of_interest$laboratory)

cell_markers_of_interest = cell_markers[cell_type_of_interest]

total_height = max(lengths(cell_markers_of_interest)) + 1

# Violin plot
patchwork_list = lapply(names(cell_markers_of_interest), FUN = function(cell_type) {
  markers_set = cell_markers_of_interest[[cell_type]]
  
  # Individual violin plot
  sub_plot_list = lapply(markers_set, FUN = function(one_gene) {
    p = Seurat::VlnPlot(subsobj_of_interest,
                        features = one_gene, pt.size = 0,
                        group.by = "cell_type",
                        split.by = "laboratory",
                        cols = laboratory_colors,
                        combine = FALSE)[[1]] +
      ggplot2::labs(y = one_gene) +
      ggplot2::scale_y_continuous(expand = c(0.01, 0)) +
      ggplot2::geom_vline(mapping = NULL, data = NULL,
                          xintercept = c(1:4) + 0.5,
                          col = "gray50", lty = 2) +
      Seurat::NoLegend() +
      ggplot2::theme(plot.title = element_blank(),
                     axis.title.x = element_blank(), # "Identity"
                     axis.line.x = element_blank(),  # replaced by panel.border
                     axis.title.y = element_text(size = 8, angle = 0, vjust = 0.5), # gene
                     axis.text.y = element_blank(), # "Expression level"
                     axis.text.x = element_blank(), # cell type
                     axis.ticks = element_blank(),
                     plot.margin = unit(c(0,0,0,0), "cm"),
                     panel.border = element_rect(size = 1, color = "gray50"))
    return(p)
  })
  
  # Reset axis.text for last plot
  # sub_plot_list[[length(sub_plot_list)]] = sub_plot_list[[length(sub_plot_list)]] +
  #   ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) # cell type
  
  # Add empty plot to arrange heights
  sub_plot_list[[length(sub_plot_list) + 1]] = patchwork::plot_spacer()
  empty_plot_height = total_height - length(markers_set)
  
  # Arrange as patchwork
  pp = patchwork::wrap_plots(sub_plot_list,
                             ncol = 1,
                             heights = c(rep(1, length(markers_set)),
                                         empty_plot_height)) +
    patchwork::plot_annotation(title = cell_type)
  
  return(pp)
})
names(patchwork_list) = names(cell_markers_of_interest)

patchwork_list
```

## Non-matrix cells

We select the datasets:

```{r non_matrix_dataset_our}
sobj_atlas = list_data$our_atlas$sobj
name2D_atlas = list_data$our_atlas$name2D
sobj = list_data$non_matrix$sobj
name2D = list_data$non_matrix$name2D
my_traj = list_data$non_matrix$my_traj

cell_type_of_interest = c("HF-SCs", "ORS", "IFE basal", "IFE granular spinous")
```

Location of cells on the whole atlas:

```{r non_matrix_location_our, fig.width = 2, fig.height = 2}
sobj_atlas$cell_bc = colnames(sobj_atlas)
sobj$cell_bc = colnames(sobj)
sobj_atlas$is_of_interest = dplyr::left_join(sobj_atlas@meta.data[, c("cell_bc", "percent.mt")],
                                             sobj@meta.data[, c("cell_bc", "cluster_type")],
                                             by = "cell_bc")[, "cluster_type"]

Seurat::DimPlot(sobj_atlas, reduction = name2D_atlas, pt.size = 0.000001,
                group.by = "is_of_interest", order = "TRUE") +
  ggplot2::scale_color_manual(values = c(color_markers[cell_type_of_interest], bg_color),
                              breaks = c(cell_type_of_interest, NA), na.value = bg_color) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank()) +
  Seurat::NoAxes() + Seurat::NoLegend()
```
* cell type annotation

```{r cell_type_non_matrix_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_non_matrix_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* pseudotime

```{r pseudotime_non_matrix_our, fig.width = 4, fig.height = 5}
Seurat::FeaturePlot(sobj, reduction = name2D, pt.size = 0.5,
                    features = "pseudotime") +
  ggplot2::scale_color_gradientn(colors = viridis::viridis(n = 100)) +
  ggplot2::lims(x = range(sobj@reductions[[name2D]]@cell.embeddings[, 1]),
                y = range(sobj@reductions[[name2D]]@cell.embeddings[, 2])) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank(),
                 legend.position = "bottom",
                 legend.direction = "horizontal") +
  Seurat::NoAxes()
```

* pseudotime with dynplot's function :

```{r pseudotime2_non_matrix_our, fig.width = 4, fig.height = 4}
dynplot::plot_dimred(trajectory = my_traj,
                     dimred = sobj[[name2D]]@cell.embeddings,
                     # Cells
                     color_cells = 'pseudotime',
                     size_cells = 0.5,
                     border_radius_percentage = 0,
                     # Trajectory
                     plot_trajectory = TRUE,
                     color_trajectory = "none",
                     label_milestones = FALSE,
                     size_milestones = 0,
                     size_transitions = 1)
```

* violin plots

```{r non_matrix_common_genes_vln, fig.width = 2.5, fig.height = 4}
features_oi = c("SOX9", "TCF4", "TBX1", "NFATC1", 
                "LHX2", "RUNX1", "VIM", "TCF3", "FGF18", "CD34",
                # HF-SCs + IFE basal
                "COL17A1", "MT2A", "TXNIP", "CAVIN1",
                # HF-SCs + ORS
                "TUBB2A", "KRT6B", "MARCKSL1", "ID4", "CYP1B1", "BARX2",
                # HF-SCs
                "LMCD1", "TCEAL2", "FRZB", "THBS1", "DIO2")

plot_list = lapply(features_oi, FUN = function(one_gene) {
  p = Seurat::VlnPlot(sobj, group.by = "cluster_type",
                      features = one_gene, pt.size = 0.001) +
    ggplot2::scale_fill_manual(values = color_markers[levels(sobj$cluster_type)],
                               breaks = levels(sobj$cluster_type)) +
    ggplot2::theme(plot.subtitle = element_text(hjust = 0.5),
                   axis.title.x = element_blank(),
                   legend.position = "none")
  return(p)
})

names(plot_list) = features_oi
plot_list
```

* feature plots

```{r non_matrix_common_genes_feature, fig.width = 4, fig.height = 4}
plot_list = lapply(features_oi, FUN = function(one_gene) {
  p = Seurat::FeaturePlot(sobj, features = one_gene, reduction = name2D) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    ggplot2::lims(x = range(sobj@reductions[[name2D]]@cell.embeddings[, 1]),
                  y = range(sobj@reductions[[name2D]]@cell.embeddings[, 2])) +
    ggplot2::theme(aspect.ratio = 1) +
    Seurat::NoAxes() +
    Seurat::NoLegend()
  
  return(p)
})

names(plot_list) = features_oi
plot_list
```

## ORS vs IFE basal

### Our dataset

We select the dataset:

```{r ors_ifeb_our}
sobj_atlas = list_data$our_atlas$sobj
name2D_atlas = list_data$our_atlas$name2D
sobj = list_data$ors_ifeb$sobj
name2D = list_data$ors_ifeb$name2D
list_results = list_data$ors_ifeb$list_results
```

Location of cells on the whole atlas:

```{r ors_ifeb_location_our, fig.width = 2, fig.height = 2}
sobj_atlas$cell_bc = colnames(sobj_atlas)
sobj$cell_bc = colnames(sobj)
sobj_atlas$is_of_interest = dplyr::left_join(sobj_atlas@meta.data[, c("cell_bc", "percent.mt")],
                                             sobj@meta.data[, c("cell_bc", "cell_type")],
                                             by = "cell_bc")[, "cell_type"]

Seurat::DimPlot(sobj_atlas, reduction = name2D_atlas, pt.size = 0.000001,
                group.by = "is_of_interest", order = "TRUE") +
  ggplot2::scale_color_manual(values = c(color_markers[cell_type_of_interest], bg_color),
                              breaks = c(cell_type_of_interest, NA), na.value = bg_color) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank()) +
  Seurat::NoAxes() + Seurat::NoLegend()
```

DE genes between ORS and IFE basal :

```{r de_ors_ifeb_our, fig.width = 6, fig.height = 6}
mark = list_results$ORS_vs_IFE_basal$mark
mark$gene_name = rownames(mark)
mark_label = rbind(
  # up-regulated in ORS
  mark %>% dplyr::top_n(., n = 20, wt = avg_logFC),
  # up-regulated in IFE basal
  mark %>% dplyr::top_n(., n = 20, wt = -avg_logFC),
  # representative and selective for ORS
  mark %>% dplyr::top_n(., n = 20, wt = (pct.1 - pct.2)),
  # representative and selective for IFE basal
  mark %>% dplyr::top_n(., n = 20, wt = -(pct.1 - pct.2))) %>%
  dplyr::distinct()
mark_label = mark_label[!grepl(rownames(mark_label), pattern = "^MT"), ]

avg_logFC_range = setNames(c(min(mark_label$avg_logFC), -1, 0, 1, max(mark_label$avg_logFC)),
                           nm = c("dodgerblue4", "dodgerblue3", "#B7B7B7", "firebrick3", "firebrick4"))


ggplot2::ggplot(mark, aes(x = pct.1, y = pct.2, col = avg_logFC)) +
  ggplot2::geom_abline(slope = 1, intercept = 0, lty = 2) +
  ggplot2::geom_point() +
  ggrepel::geom_label_repel(data = mark_label, max.overlaps = Inf,
                            aes(x = pct.1, y = pct.2, label = gene_name),
                            col = "black", fill = NA, size = 3.5, label.size = NA) +
  ggplot2::labs(x = "Enriched in ORS",
                y = "Enriched in IFE basal") +
  ggplot2::scale_color_gradientn(colors = names(avg_logFC_range),
                                 values = scales::rescale(unname(avg_logFC_range))) +
  ggplot2::theme_classic() +
  ggplot2::theme(aspect.ratio = 1)
```

* GSEA, enriched in ORS compared to IFE basal

```{r kera_ors_ifeb_our, fig.width = 6, fig.height = 4}
the_gs_name = "REACTOME_KERATINIZATION" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```

* GSEA, enriched in IFE basal compared to ORS

```{r ifn_ors_ifeb_our, fig.width = 6, fig.height = 4}
the_gs_name = "HALLMARK_INTERFERON_GAMMA_RESPONSE" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```


* violin plots of DE genes in ORS, between HS and HD

```{r ors_ifeb_vln_ors_our, fig.width = 2, fig.height = 3}
subsobj = subset(sobj, cell_type == "ORS")
features_oi = c("DUSP1", "DDIT4", "GABARAP", "ZNF302",
                "MIF", "LGALS7", "ARF5", "S100A9")

# Plot !
plot_list = lapply(features_oi, FUN = function(feature) {
  # t-test between sample type
  feature_expr = subsobj@assays$RNA@data[feature, ]
  feature_hs = feature_expr[subsobj$sample_type == "HS"]
  feature_hd = feature_expr[subsobj$sample_type == "HD"]
  feature_hs_VS_feature_hd = stats::t.test(feature_hs, feature_hd)
  feature_hs_VS_feature_hd
  
  # Significance associated with p-value
  pval = feature_hs_VS_feature_hd$p.value
  
  significance = case_when(pval > 0.05 ~ "ns",
                           pval > 0.01 & pval <= 0.05 ~ "*",
                           pval > 0.001 & pval <= 0.01 ~ "**",
                           pval <= 0.001 ~ "***")
  
  # Plot
  p = Seurat::VlnPlot(subsobj, group.by = "sample_type",
                      pt.size = 0.3, features = feature) +
    ggplot2::scale_fill_manual(values = sample_type_colors,
                               breaks = names(sample_type_colors)) +
    # Significance bar
    ggplot2::geom_segment(data = data.frame(1), inherit.aes = FALSE,
                          size = 0.5,
                          x = 1, xend = 2,
                          y = max(feature_expr)+0.3, yend = max(feature_expr)+0.3) +
    ggplot2::geom_label(data = data.frame(1), inherit.aes = FALSE,
                        x = 1.5, y = max(feature_expr)+0.35,
                        label = significance,
                        fill = NA, label.size = 0) +
    ggplot2::lims(y = c(0, max(feature_expr)+0.4)) +
    # Theme
    ggplot2::theme(axis.title.x = element_blank(),
                   legend.position = "none")
  
  return(p)
})

names(plot_list) = features_oi
plot_list
```

* violin plots of DE genes in IFE basal, between HS and HD

```{r ors_ifeb_vln_ifeb_our, fig.width = 2, fig.height = 3}
subsobj = subset(sobj, cell_type == "IFE basal")
features_oi = c("DUSP1", "KLF6", "CLDN1", "CTGF",
                "IFI27", "IFITM3", "CCL2", "S100A9")

# Plot !
plot_list = lapply(features_oi, FUN = function(feature) {
  # t-test between sample type
  feature_expr = subsobj@assays$RNA@data[feature, ]
  feature_hs = feature_expr[subsobj$sample_type == "HS"]
  feature_hd = feature_expr[subsobj$sample_type == "HD"]
  feature_hs_VS_feature_hd = stats::t.test(feature_hs, feature_hd)
  feature_hs_VS_feature_hd
  
  # Significance associated with p-value
  pval = feature_hs_VS_feature_hd$p.value
  
  significance = case_when(pval > 0.05 ~ "ns",
                           pval > 0.01 & pval <= 0.05 ~ "*",
                           pval > 0.001 & pval <= 0.01 ~ "**",
                           pval <= 0.001 ~ "***")
  
  # Plot
  p = Seurat::VlnPlot(subsobj, group.by = "sample_type",
                      pt.size = 0.3, features = feature) +
    ggplot2::scale_fill_manual(values = sample_type_colors,
                               breaks = names(sample_type_colors)) +
    # Significance bar
    ggplot2::geom_segment(data = data.frame(1), inherit.aes = FALSE,
                          size = 0.5,
                          x = 1, xend = 2,
                          y = max(feature_expr)+0.3, yend = max(feature_expr)+0.3) +
    ggplot2::geom_label(data = data.frame(1), inherit.aes = FALSE,
                        x = 1.5, y = max(feature_expr)+0.35,
                        label = significance,
                        fill = NA, label.size = 0) +
    ggplot2::lims(y = c(0, max(feature_expr)+0.4)) +
    # Theme
    ggplot2::theme(axis.title.x = element_blank(),
                   legend.position = "none")
  
  return(p)
})

names(plot_list) = features_oi
plot_list
```


### Wu dataset

We select the dataset:

```{r ors_ifeb_wu}
sobj_atlas = list_data$wu_atlas$sobj
name2D_atlas = list_data$wu_atlas$name2D
sobj = list_data$wu_ors_ifeb$sobj
name2D = list_data$wu_ors_ifeb$name2D
list_results = list_data$wu_ors_ifeb$list_results
```

Location of cells on the whole atlas:

```{r ors_ifeb_location_wu, fig.width = 2, fig.height = 2}
sobj_atlas$cell_bc = colnames(sobj_atlas)
sobj$cell_bc = colnames(sobj)
sobj_atlas$is_of_interest = dplyr::left_join(sobj_atlas@meta.data[, c("cell_bc", "percent.mt")],
                                             sobj@meta.data[, c("cell_bc", "cell_type")],
                                             by = "cell_bc")[, "cell_type"]

Seurat::DimPlot(sobj_atlas, reduction = name2D_atlas, pt.size = 0.000001,
                group.by = "is_of_interest", order = "TRUE") +
  ggplot2::scale_color_manual(values = c(color_markers[cell_type_of_interest], bg_color),
                              breaks = c(cell_type_of_interest, NA), na.value = bg_color) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank()) +
  Seurat::NoAxes() + Seurat::NoLegend()
```

* GSEA, enriched in ORS compared to IFE basal

```{r kera_ors_ifeb_wu, fig.width = 6, fig.height = 4}
the_gs_name = "REACTOME_KERATINIZATION" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```

* GSEA, enriched in IFE basal compared to ORS

```{r ifn_ors_ifeb_wu, fig.width = 6, fig.height = 4}
the_gs_name = "HALLMARK_INTERFERON_GAMMA_RESPONSE" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```

### Takahashi dataset

We select the dataset:

```{r ors_ifeb_takahashi}
sobj_atlas = list_data$takahashi_atlas$sobj
name2D_atlas = list_data$takahashi_atlas$name2D
sobj = list_data$takahashi_ors_ifeb$sobj
name2D = list_data$takahashi_ors_ifeb$name2D
list_results = list_data$takahashi_ors_ifeb$list_results
```

Location of cells on the whole atlas:

```{r ors_ifeb_location_takahashi, fig.width = 2, fig.height = 2}
sobj_atlas$cell_bc = colnames(sobj_atlas)
sobj$cell_bc = colnames(sobj)
sobj_atlas$is_of_interest = dplyr::left_join(sobj_atlas@meta.data[, c("cell_bc", "percent.mt")],
                                             sobj@meta.data[, c("cell_bc", "cell_type")],
                                             by = "cell_bc")[, "cell_type"]

Seurat::DimPlot(sobj_atlas, reduction = name2D_atlas, pt.size = 0.000001,
                group.by = "is_of_interest", order = "TRUE") +
  ggplot2::scale_color_manual(values = c(color_markers[cell_type_of_interest], bg_color),
                              breaks = c(cell_type_of_interest, NA), na.value = bg_color) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank()) +
  Seurat::NoAxes() + Seurat::NoLegend()
```

* GSEA, enriched in ORS compared to IFE basal

```{r kera_ors_ifeb_takahashi, fig.width = 6, fig.height = 4}
the_gs_name = "REACTOME_KERATINIZATION" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```

* GSEA, enriched in IFE basal compared to ORS

```{r ifn_ors_ifeb_takahashi, fig.width = 6, fig.height = 4}
the_gs_name = "HALLMARK_INTERFERON_GAMMA_RESPONSE" 
the_content = list_results$ORS_vs_IFE_basal$gsea@result %>%
  dplyr::filter(ID == the_gs_name)
the_subtitle = paste0("\nNES : ", round(the_content$NES, 2),
                      "\t|\tpvalue : ", round(the_content$pvalue, 4),
                      "\t|\tset size : ", the_content$setSize, " genes")

enrichplot::gseaplot2(x = list_results$ORS_vs_IFE_basal$gsea,
                      geneSetID = the_gs_name) +
  ggplot2::labs(title = the_gs_name,
                subtitle = the_subtitle) +
  ggplot2::theme(plot.title = element_text(hjust = 0.5, face = "bold",
                                           margin = ggplot2::margin(3, 3, 5, 3)),
                 plot.subtitle = ggtext::element_markdown(hjust = 0.5,
                                                          size = 10))
```

## HF-SCs

We select the dataset:

```{r hfscs_our}
sobj = list_data$hfsc$sobj
name2D = list_data$hfsc$name2D
list_results = list_data$hfsc$list_results
```

* violin plots of DE genes in HF-SCs, between HS and HD

```{r hfscs_vln_ors_our, fig.width = 2, fig.height = 3}
features_oi = c("HLA-C", "HLA-A", "IFITM3", "MIF", "JUNB", "PPIA",
                "PDK4", "DDIT4", "BTG2", "TXNIP", "KLF9")

# Plot !
plot_list = lapply(features_oi, FUN = function(feature) {
  # t-test between sample type
  feature_expr = sobj@assays$RNA@data[feature, ]
  feature_hs = feature_expr[sobj$sample_type == "HS"]
  feature_hd = feature_expr[sobj$sample_type == "HD"]
  feature_hs_VS_feature_hd = stats::t.test(feature_hs, feature_hd)
  feature_hs_VS_feature_hd
  
  # Significance associated with p-value
  pval = feature_hs_VS_feature_hd$p.value
  
  significance = case_when(pval > 0.05 ~ "ns",
                           pval > 0.01 & pval <= 0.05 ~ "*",
                           pval > 0.001 & pval <= 0.01 ~ "**",
                           pval <= 0.001 ~ "***")
  
  # Plot
  p = Seurat::VlnPlot(sobj, group.by = "sample_type",
                      pt.size = 0.3, features = feature) +
    ggplot2::scale_fill_manual(values = sample_type_colors,
                               breaks = names(sample_type_colors)) +
    # Significance bar
    ggplot2::geom_segment(data = data.frame(1), inherit.aes = FALSE,
                          size = 0.5,
                          x = 1, xend = 2,
                          y = max(feature_expr)+0.3, yend = max(feature_expr)+0.3) +
    ggplot2::geom_label(data = data.frame(1), inherit.aes = FALSE,
                        x = 1.5, y = max(feature_expr)+0.35,
                        label = significance,
                        fill = NA, label.size = 0) +
    ggplot2::lims(y = c(0, max(feature_expr)+0.4)) +
    # Theme
    ggplot2::theme(axis.title.x = element_blank(),
                   legend.position = "none")
  
  return(p)
})

names(plot_list) = features_oi
plot_list
```
* heatmap

Heatmap between HS and HD :

```{r hfscs_heatmap_our, fig.width = 5.5, fig.height = 10}
# features_oi = list_results$HS_vs_HD %>%
#   dplyr::filter(pct.1 > 0.5 | pct.2 > 0.5) %>%
#   dplyr::arrange(-avg_logFC, -pct.1, -pct.2) %>%
#   rownames()
# features_oi = features_oi[!grepl(features_oi, pattern = "^RP")]

features_oi = c("IFITM3", "MIF", "HLA-C", "LMCD1", "TGFB2", "CYP1B1",
                "KRT6B", "WIF1", "HSPA1A", "COMT", "CISD1", "PRNP",  "CD81",
                "HLA-A", "B2M", "PPIA", "PYCARD", "CSAD", "PFN1", "HIF1A",
                "BASP1", "CAV1", "ZNF302", "LMO4", "SEPT11", "HTRA1",
                "GPM6A", "SOX4", "ZFP36L2", "RHOB", "CLDN1", "DUSP1", "TBX1",
                "ATF3", "DDIT4", "TXNIP", "BTG2", "SGK1", "KLF6", "LGR5")

# Matrix
mat_expr = Seurat::GetAssayData(sobj)
mat_expr = mat_expr[features_oi, ]
mat_expr = Matrix::t(mat_expr)
mat_expr = cbind(mat_expr, sobj$percent.mt)
colnames(mat_expr)[ncol(mat_expr)] = "percent.rb"
mat_expr = dynutils::scale_quantile(mat_expr) # between 0 and 1
mat_expr = Matrix::t(mat_expr)
dim(mat_expr) # genes x cells

## Colors
list_colors = list()

# Heatmap
list_colors[["expression"]] = rev(RColorBrewer::brewer.pal(name = "RdBu", n = 9))

# Sample annotation (top annotation)
list_colors[["sample_type"]] = sample_type_colors
list_colors[["sample_identifier"]] = setNames(nm = sample_info$sample_identifier,
                                              sample_info$color)
# list_colors[["seurat_clusters"]] = setNames(aquarius::gg_color_hue(length(levels(sobj$seurat_clusters))),
#                                             nm = levels(sobj$seurat_clusters))
# Cells order
column_order = sobj@meta.data %>%
  dplyr::arrange(sample_type, sample_identifier) %>%
  rownames()
column_order = match(column_order, rownames(sobj@meta.data))

# Heatmap
ha_top = HeatmapAnnotation(sample_type = sobj$sample_type,
                           sample_identifier = sobj$sample_identifier,
                           # clusters = sobj$seurat_clusters,
                           col = list(sample_type = list_colors[["sample_type"]],
                                      sample_identifier = list_colors[["sample_identifier"]]
                                      # clusters = list_colors[["seurat_clusters"]]
                           ))


# g1 : REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM
# g2 : GOBP_APOPTOTIC_PROCESS
g1_genes = c("B2M", "HLA-C", "HLA-A", "MIF", "PPIA", "JUNB", "IFITM3")
g2_genes = c("JUN", "ATF3", "BTG2", "RHOB", "NFKBIA", "SGK1", "KLF9",
             "CAV1", "DDIT4", "PDK4", "TXNIP", "RNF1152", "TLE1")
ha_right = data.frame(genes =  c(features_oi, "percent.rb"), rownames = c(features_oi, "percent.rb"))
ha_right$group = case_when(ha_right$genes %in% g1_genes ~ "REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM",
                           ha_right$genes %in% g2_genes ~ "GOBP_APOPTOTIC_PROCESS",
                           TRUE ~ "others")

list_colors[["group"]] = setNames(
  nm = c("REACTOME_CYTOKINE_SIGNALING_IN_IMMUNE_SYSTEM", "GOBP_APOPTOTIC_PROCESS", "others"),
  c("red", "black", "gray90"))

ha_right = HeatmapAnnotation(group = ha_right$group,
                             col = list(group = list_colors[["group"]]),
                             which = "row",
                             show_annotation_name = FALSE,
                             show_legend = TRUE)

# Heatmap
ht = Heatmap(as.matrix(mat_expr),
             heatmap_legend_param = list(title = "Expression", at = c(0, 1), 
                                         labels = c("low", "high")),
             col = list_colors[["expression"]],
             top_annotation = ha_top,
             right_annotation = ha_right,
             show_column_names = FALSE,
             column_order = column_order,
             column_gap = unit(2, "mm"),
             cluster_rows = FALSE,
             show_row_dend = FALSE,
             row_title = NULL,
             row_names_gp = grid::gpar(fontsize = 10, fontface = "plain"),
             use_raster = FALSE,
             show_heatmap_legend = TRUE,
             border = TRUE)

ComplexHeatmap::draw(ht,
                     merge_legend = TRUE,
                     heatmap_legend_side = "bottom",
                     annotation_legend_side = "bottom")
```

## Immune cells

We select the datasets:

```{r immune_dataset_our}
sobj_atlas = list_data$our_atlas$sobj
name2D_atlas = list_data$our_atlas$name2D
sobj = list_data$immune$sobj
name2D = list_data$immune$name2D

cell_type_of_interest = c("CD4 T cells", "CD8 T cells", "Langerhans cells",
                          "macrophages", "B cells", "proliferative")
```

Location of cells on the whole atlas:

```{r immune_location_our, fig.width = 2, fig.height = 2}
sobj_atlas$cell_bc = colnames(sobj_atlas)
sobj$cell_bc = colnames(sobj)
sobj_atlas$is_of_interest = dplyr::left_join(sobj_atlas@meta.data[, c("cell_bc", "percent.mt")],
                                             sobj@meta.data[, c("cell_bc", "cluster_type")],
                                             by = "cell_bc")[, "cluster_type"]

Seurat::DimPlot(sobj_atlas, reduction = name2D_atlas, pt.size = 0.000001,
                group.by = "is_of_interest", order = "TRUE") +
  ggplot2::scale_color_manual(values = c(color_markers[cell_type_of_interest], bg_color),
                              breaks = c(cell_type_of_interest, NA), na.value = bg_color) +
  ggplot2::theme(aspect.ratio = 1,
                 plot.title = element_blank()) +
  Seurat::NoAxes() + Seurat::NoLegend()
```

* cell type annotation

```{r cell_type_immune_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cell_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* cluster type annotation

```{r cluster_type_immune_our, fig.width = 4, fig.height = 4}
Seurat::DimPlot(sobj, reduction = name2D, pt.size = 0.5,
                group.by = "cluster_type", cols = color_markers) +
  ggplot2::theme(aspect.ratio = 1) +
  Seurat::NoAxes() +
  Seurat::NoLegend()
```

* feature plots for genes related to TCR alpha

```{r gene_immune_our, fig.width = 12, fig.height = 8}
features_oi = c("CD3E", "CD4", "CD8A",
                sort(grep(rownames(sobj), pattern = "^TRA[C|V]", value = TRUE)))

plot_list = lapply(features_oi, FUN = function(one_gene) {
  p = Seurat::FeaturePlot(sobj, features = one_gene,
                          reduction = name2D, order = TRUE) +
    ggplot2::scale_color_gradientn(colors = aquarius::color_gene) +
    ggplot2::theme(aspect.ratio = 1) +
    Seurat::NoAxes() +
    Seurat::NoLegend()
  
  return(p)
})

patchwork::wrap_plots(plot_list, nrow = 4)
```

* feature plots split by sample type (HS / HD)

```{r gene_split, immune_our, fig.width = 4, fig.height = 2}
features_oi = data.frame(population = c("all immune cells", rep("macrophages", 2),
                                      rep("CD4 T cells", 3), rep("CD8 T cells", 2)),
                         feature = c("IL6", "IL1B", "TNF", "GZMA", "IFNG", "IL17A", "PRF1", "GZMB"))
features_oi

patchwork_list = lapply(features_oi$feature, FUN = function(one_gene) {
  plot_list = aquarius::plot_split_dimred(sobj,
                                          reduction = name2D,
                                          split_by = "sample_type",
                                          color_by = one_gene,
                                          order = TRUE,
                                          bg_color = bg_color)
  
  p = patchwork::wrap_plots(plot_list, nrow = 1) +
    patchwork::plot_layout(guides = "collect")
  
  return(p)
})
names(patchwork_list) = features_oi$feature

patchwork_list
```

* violin plots within a specific population

```{r gene_split_immune_our, fig.width = 2, fig.height = 3}
plot_list = lapply(c(1:nrow(features_oi)), FUN = function(i) {
  one_gene = features_oi$feature[i]
  population = features_oi$population[i]
  
  # Subset object
  if (population != "all immune cells") {
    subsobj = subset(sobj, cluster_type %in% population)
  } else {
   subsobj = sobj 
  }
  
  # t-test between sample type
  feature_expr = subsobj@assays$RNA@data[one_gene, ]
  feature_hs = feature_expr[subsobj$sample_type == "HS"]
  feature_hd = feature_expr[subsobj$sample_type == "HD"]
  feature_hs_VS_feature_hd = stats::t.test(feature_hs, feature_hd)
  feature_hs_VS_feature_hd
  
  # Significance associated with p-value
  pval = feature_hs_VS_feature_hd$p.value
  
  significance = case_when(pval > 0.05 ~ "ns",
                           pval > 0.01 & pval <= 0.05 ~ "*",
                           pval > 0.001 & pval <= 0.01 ~ "**",
                           pval <= 0.001 ~ "***")
  
  # Plot
  p = Seurat::VlnPlot(subsobj, group.by = "sample_type",
                      pt.size = 0.3, features = one_gene) +
    ggplot2::scale_fill_manual(values = sample_type_colors,
                               breaks = names(sample_type_colors)) +
    ggplot2::labs(title = population,
                  subtitle = one_gene) +
    # Significance bar
    ggplot2::geom_segment(data = data.frame(1), inherit.aes = FALSE,
                          size = 0.5,
                          x = 1, xend = 2,
                          y = max(feature_expr)+0.3, yend = max(feature_expr)+0.3) +
    ggplot2::geom_label(data = data.frame(1), inherit.aes = FALSE,
                        x = 1.5, y = max(feature_expr)+0.35,
                        label = significance,
                        fill = NA, label.size = 0) +
    ggplot2::lims(y = c(0, max(feature_expr)+0.4)) +
    # Theme
    ggplot2::theme(axis.title.x = element_blank(),
                   plot.title = element_text(size = 10),
                   plot.subtitle = element_text(hjust = 0.5),
                   legend.position = "none")
  
  return(p)
})
names(plot_list) = features_oi$feature

plot_list
```


















* heatmap macrophages

```{r heatmap_immune_macrophages, fig.width = 5, fig.height = 5.2}
subsobj = subset(sobj, cluster_type == "macrophages")
features_oi = c("IL1B", "TNF",
                "HLA-DQA2", "HLA-DPA1", "HLA-DRB5",
                "HLA-A", "HLA-C", "B2M",
                "C1QA", "C1QB", "C1QC")

# Matrix
mat_expr = Seurat::GetAssayData(subsobj)
mat_expr = mat_expr[features_oi, ]
mat_expr = Matrix::t(mat_expr)
mat_expr = dynutils::scale_quantile(mat_expr) # between 0 and 1
mat_expr = Matrix::t(mat_expr)
dim(mat_expr) # genes x cells
## Colors
list_colors = list()

# Heatmap
list_colors[["expression"]] = rev(RColorBrewer::brewer.pal(name = "RdBu", n = 9))

# Sample annotation (top annotation)
list_colors[["sample_type"]] = sample_type_colors
list_colors[["sample_identifier"]] = setNames(nm = sample_info$sample_identifier,
                                              sample_info$color)
# Cells order
column_order = subsobj@meta.data %>%
  dplyr::arrange(sample_type, sample_identifier) %>%
  rownames()
column_order = match(column_order, rownames(subsobj@meta.data))

# Heatmap
ha_top = HeatmapAnnotation(sample_type = subsobj$sample_type,
                           sample_identifier = subsobj$sample_identifier,
                           col = list(sample_type = list_colors[["sample_type"]],
                                      sample_identifier = list_colors[["sample_identifier"]]))

# Heatmap
ht = Heatmap(as.matrix(mat_expr),
             heatmap_legend_param = list(title = "Expression", at = c(0, 1), 
                                         labels = c("low", "high")),
             col = list_colors[["expression"]],
             top_annotation = ha_top,
             show_column_names = FALSE,
             column_order = column_order,
             column_gap = unit(2, "mm"),
             cluster_rows = FALSE,
             row_title = NULL,
             row_names_gp = grid::gpar(fontsize = 14, fontface = "plain"),
             use_raster = FALSE,
             show_heatmap_legend = TRUE,
             border = TRUE)

ComplexHeatmap::draw(ht,
                     merge_legend = TRUE,
                     heatmap_legend_side = "bottom",
                     annotation_legend_side = "bottom")
```

* heatmap T cells

```{r heatmap_immune_t_cells, fig.width = 6, fig.height = 4}
subsobj = subset(sobj, cluster_type == "CD4 T cells")
features_oi = c("GZMA", "KLRB1", "BTG1", "ZFP36", "NFKBIA", "TXNIP", "CXCR4", "IFNG", "IL17A")

# Matrix
mat_expr = Seurat::GetAssayData(subsobj)
mat_expr = mat_expr[features_oi, ]
mat_expr = Matrix::t(mat_expr)
mat_expr = dynutils::scale_quantile(mat_expr) # between 0 and 1
mat_expr = Matrix::t(mat_expr)
dim(mat_expr) # genes x cells
## Colors
list_colors = list()

# Heatmap
list_colors[["expression"]] = rev(RColorBrewer::brewer.pal(name = "RdBu", n = 9))

# Sample annotation (top annotation)
list_colors[["sample_type"]] = sample_type_colors
list_colors[["sample_identifier"]] = setNames(nm = sample_info$sample_identifier,
                                              sample_info$color)
# Cells order
column_order = subsobj@meta.data %>%
  dplyr::arrange(sample_type, sample_identifier) %>%
  rownames()
column_order = match(column_order, rownames(subsobj@meta.data))

# Heatmap
ha_top = HeatmapAnnotation(sample_type = subsobj$sample_type,
                           sample_identifier = subsobj$sample_identifier,
                           col = list(sample_type = list_colors[["sample_type"]],
                                      sample_identifier = list_colors[["sample_identifier"]]))

# Heatmap
ht = Heatmap(as.matrix(mat_expr),
             heatmap_legend_param = list(title = "Expression", at = c(0, 1), 
                                         labels = c("low", "high")),
             col = list_colors[["expression"]],
             top_annotation = ha_top,
             show_column_names = FALSE,
             column_order = column_order,
             column_gap = unit(2, "mm"),
             cluster_rows = FALSE,
             row_title = NULL,
             row_names_gp = grid::gpar(fontsize = 14, fontface = "plain"),
             use_raster = FALSE,
             show_heatmap_legend = TRUE,
             border = TRUE)

ComplexHeatmap::draw(ht,
                     merge_legend = TRUE,
                     heatmap_legend_side = "left",
                     annotation_legend_side = "left")
```
















# Supplementary tables

In this section, we save files to associate with the manuscript, as supplementary tables.

## Table S2 (package version)

We load the table :

```{r ts2_load}
package_version = read.table(paste0(".", "/data/info_to_install_2023_04_17.txt"),
                             header = TRUE)
head(package_version)
```

Columns correspond to :

* **order** : order to install package such that one never need to install dependencies
* **package_name** : package name
* **version** : installed version, on the local machine
* **url** : the package was installed in the Singularity container using this url

We save this table, except the last column (just for me) :

```{r ts2_save}
openxlsx::write.xlsx(package_version[, c(1:4)],
                     file = paste0(".", "/data/Supplementary Table 2.xlsx"))
```

## Table S3 (cell type annotation)

We load the table :

```{r ts3_load}
cell_markers = readRDS(paste0(data_dir, "/1_metadata/hs_hd_cell_markers.rds"))
lengths(cell_markers)
```

We save this table, except the last column (just for me) :

```{r ts3_save}
openxlsx::write.xlsx(cell_markers,
                     file = paste0(".", "/data/Supplementary Table 3.xlsx"))
```

# R Session

```{r sessioninfo, echo = FALSE, fold_output = TRUE}
sessionInfo()
```
